// angular-auth-oidc-client/modules/angular-auth-oidc-client.js
import { EventEmitter, Inject, Injectable, NgModule, NgZone, Output, PLATFORM_ID } from '@angular/core';
import { BehaviorSubject, Observable, throwError } from 'rxjs';
import { isPlatformBrowser } from '@angular/common';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Router } from '@angular/router';
import { catchError, map } from 'rxjs/operators';
import { KEYUTIL, KJUR, hextob64u } from 'jsrsasign';

var __awaiter = (undefined && undefined.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcConfigService {
    constructor() {
        this.onConfigurationLoaded = new EventEmitter();
    }
    /**
     * @param {?} configUrl
     * @return {?}
     */
    load(configUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const /** @type {?} */ response = yield fetch(configUrl);
                if (!response.ok) {
                    throw new Error(response.statusText);
                }
                this.clientConfiguration = yield response.json();
                yield this.load_using_stsServer(this.clientConfiguration.stsServer);
            }
            catch (/** @type {?} */ err) {
                console.error(`OidcConfigService 'load' threw an error on calling ${configUrl}`, err);
                this.onConfigurationLoaded.emit(false);
            }
        });
    }
    /**
     * @param {?} stsServer
     * @return {?}
     */
    load_using_stsServer(stsServer) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const /** @type {?} */ response = yield fetch(`${stsServer}/.well-known/openid-configuration`);
                if (!response.ok) {
                    throw new Error(response.statusText);
                }
                this.wellKnownEndpoints = yield response.json();
                this.onConfigurationLoaded.emit(true);
            }
            catch (/** @type {?} */ err) {
                console.error(`OidcConfigService 'load_using_stsServer' threw an error on calling ${stsServer}`, err);
                this.onConfigurationLoaded.emit(false);
            }
        });
    }
    /**
     * @param {?} stsServer
     * @return {?}
     */
    load_using_custom_stsServer(stsServer) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const /** @type {?} */ response = yield fetch(stsServer);
                if (!response.ok) {
                    throw new Error(response.statusText);
                }
                this.wellKnownEndpoints = yield response.json();
                this.onConfigurationLoaded.emit(true);
            }
            catch (/** @type {?} */ err) {
                console.error(`OidcConfigService 'load_using_custom_stsServer' threw an error on calling ${stsServer}`, err);
                this.onConfigurationLoaded.emit(false);
            }
        });
    }
}
OidcConfigService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcConfigService.propDecorators = {
    "onConfigurationLoaded": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {number} */
const AuthorizationResult = {
    authorized: 1,
    forbidden: 2,
    unauthorized: 3,
};
AuthorizationResult[AuthorizationResult.authorized] = "authorized";
AuthorizationResult[AuthorizationResult.forbidden] = "forbidden";
AuthorizationResult[AuthorizationResult.unauthorized] = "unauthorized";

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class ValidateStateResult {
    /**
     * @param {?=} access_token
     * @param {?=} id_token
     * @param {?=} authResponseIsValid
     * @param {?=} decoded_id_token
     */
    constructor(access_token = '', id_token = '', authResponseIsValid = false, decoded_id_token) {
        this.access_token = access_token;
        this.id_token = id_token;
        this.authResponseIsValid = authResponseIsValid;
        this.decoded_id_token = decoded_id_token;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class DefaultConfiguration {
    constructor() {
        this.stsServer = 'https://localhost:44318';
        this.redirect_url = 'https://localhost:44311';
        // The Client MUST validate that the aud (audience) Claim contains its client_id value registered at the Issuer identified
        // by the iss (issuer) Claim as an audience.
        // The ID Token MUST be rejected if the ID Token does not list the Client as a valid audience,
        // or if it contains additional audiences not trusted by the Client.
        this.client_id = 'angularclient';
        this.response_type = 'id_token token';
        this.scope = 'openid email profile';
        // Only for Google Auth with particular G Suite domain, see https://developers.google.com/identity/protocols/OpenIDConnect#hd-param
        this.hd_param = '';
        this.post_logout_redirect_uri = 'https://localhost:44311/unauthorized';
        this.start_checksession = false;
        this.silent_renew = true;
        this.silent_renew_url = 'https://localhost:44311';
        this.silent_renew_offset_in_seconds = 0;
        this.silent_redirect_url = 'https://localhost:44311';
        this.post_login_route = '/';
        // HTTP 403
        this.forbidden_route = '/forbidden';
        // HTTP 401
        this.unauthorized_route = '/unauthorized';
        this.auto_userinfo = true;
        this.auto_clean_state_after_authentication = true;
        this.trigger_authorization_result_event = false;
        this.log_console_warning_active = true;
        this.log_console_debug_active = false;
        // id_token C8: The iat Claim can be used to reject tokens that were issued too far away from the current time,
        // limiting the amount of time that nonces need to be stored to prevent attacks.The acceptable range is Client specific.
        this.max_id_token_iat_offset_allowed_in_seconds = 3;
        this.storage = typeof Storage !== 'undefined' ? sessionStorage : null;
    }
}
class OpenIDImplicitFlowConfiguration {
    constructor() {
        this.stsServer = 'https://localhost:44318';
        this.redirect_url = 'https://localhost:44311';
        this.client_id = 'angularclient';
        this.response_type = 'id_token token';
        this.resource = '';
        this.scope = 'openid email profile';
        this.hd_param = '';
        this.post_logout_redirect_uri = 'https://localhost:44311/unauthorized';
        this.start_checksession = false;
        this.silent_renew = true;
        this.silent_renew_url = 'https://localhost:44311';
        this.silent_renew_offset_in_seconds = 0;
        this.silent_redirect_url = 'https://localhost:44311';
        this.post_login_route = '/';
        this.forbidden_route = '/forbidden';
        this.unauthorized_route = '/unauthorized';
        this.auto_userinfo = true;
        this.auto_clean_state_after_authentication = true;
        this.trigger_authorization_result_event = false;
        this.log_console_warning_active = true;
        this.log_console_debug_active = false;
        this.max_id_token_iat_offset_allowed_in_seconds = 3;
        this.storage = sessionStorage;
    }
}
class AuthConfiguration {
    /**
     * @param {?} defaultConfig
     */
    constructor(defaultConfig) {
        this.defaultConfig = defaultConfig;
    }
    /**
     * @return {?}
     */
    get stsServer() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.stsServer;
        }
        return this.defaultConfig.stsServer;
    }
    /**
     * @return {?}
     */
    get redirect_url() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.redirect_url;
        }
        return this.defaultConfig.redirect_url;
    }
    /**
     * @return {?}
     */
    get silent_redirect_url() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.silent_renew_url;
        }
        return this.defaultConfig.silent_renew_url;
    }
    /**
     * @return {?}
     */
    get client_id() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.client_id;
        }
        return this.defaultConfig.client_id;
    }
    /**
     * @return {?}
     */
    get response_type() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.response_type;
        }
        return this.defaultConfig.response_type;
    }
    /**
     * @return {?}
     */
    get scope() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.scope;
        }
        return this.defaultConfig.scope;
    }
    /**
     * @return {?}
     */
    get hd_param() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.hd_param;
        }
        return this.defaultConfig.hd_param;
    }
    /**
     * @return {?}
     */
    get post_logout_redirect_uri() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.post_logout_redirect_uri;
        }
        return this.defaultConfig.post_logout_redirect_uri;
    }
    /**
     * @return {?}
     */
    get start_checksession() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.start_checksession;
        }
        return this.defaultConfig.start_checksession;
    }
    /**
     * @return {?}
     */
    get silent_renew() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.silent_renew;
        }
        return this.defaultConfig.silent_renew;
    }
    /**
     * @return {?}
     */
    get silent_renew_offset_in_seconds() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.silent_renew_offset_in_seconds;
        }
        return this.defaultConfig.silent_renew_offset_in_seconds;
    }
    /**
     * @return {?}
     */
    get post_login_route() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.post_login_route;
        }
        return this.defaultConfig.post_login_route;
    }
    /**
     * @return {?}
     */
    get forbidden_route() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.forbidden_route;
        }
        return this.defaultConfig.forbidden_route;
    }
    /**
     * @return {?}
     */
    get unauthorized_route() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.unauthorized_route;
        }
        return this.defaultConfig.unauthorized_route;
    }
    /**
     * @return {?}
     */
    get auto_userinfo() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.auto_userinfo;
        }
        return this.defaultConfig.auto_userinfo;
    }
    /**
     * @return {?}
     */
    get auto_clean_state_after_authentication() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.auto_clean_state_after_authentication;
        }
        return this.defaultConfig.auto_clean_state_after_authentication;
    }
    /**
     * @return {?}
     */
    get trigger_authorization_result_event() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.trigger_authorization_result_event;
        }
        return this.defaultConfig.trigger_authorization_result_event;
    }
    /**
     * @return {?}
     */
    get log_console_warning_active() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.log_console_warning_active;
        }
        return this.defaultConfig.log_console_warning_active;
    }
    /**
     * @return {?}
     */
    get log_console_debug_active() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.log_console_debug_active;
        }
        return this.defaultConfig.log_console_debug_active;
    }
    /**
     * @return {?}
     */
    get max_id_token_iat_offset_allowed_in_seconds() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.max_id_token_iat_offset_allowed_in_seconds;
        }
        return this.defaultConfig.max_id_token_iat_offset_allowed_in_seconds;
    }
    /**
     * @return {?}
     */
    get storage() {
        if (this.openIDImplicitFlowConfiguration) {
            return this.openIDImplicitFlowConfiguration.storage;
        }
        return this.defaultConfig.storage;
    }
    /**
     * @param {?} openIDImplicitFlowConfiguration
     * @return {?}
     */
    init(openIDImplicitFlowConfiguration) {
        this.openIDImplicitFlowConfiguration = openIDImplicitFlowConfiguration;
    }
}
AuthConfiguration.decorators = [
    { type: Injectable },
];
/** @nocollapse */
AuthConfiguration.ctorParameters = () => [
    { type: DefaultConfiguration, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Implement this class-interface to create a custom storage.
 * @abstract
 */
class OidcSecurityStorage {
}
OidcSecurityStorage.decorators = [
    { type: Injectable },
];
class BrowserStorage {
    /**
     * @param {?} authConfiguration
     */
    constructor(authConfiguration) {
        this.authConfiguration = authConfiguration;
        this.hasStorage = typeof Storage !== 'undefined';
    }
    /**
     * @param {?} key
     * @return {?}
     */
    read(key) {
        if (this.hasStorage) {
            return JSON.parse(this.authConfiguration.storage.getItem(key));
        }
        return;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    write(key, value) {
        if (this.hasStorage) {
            value = value === undefined ? null : value;
            this.authConfiguration.storage.setItem(key, JSON.stringify(value));
        }
    }
}
BrowserStorage.decorators = [
    { type: Injectable },
];
/** @nocollapse */
BrowserStorage.ctorParameters = () => [
    { type: AuthConfiguration, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcSecurityCommon {
    /**
     * @param {?} oidcSecurityStorage
     */
    constructor(oidcSecurityStorage) {
        this.oidcSecurityStorage = oidcSecurityStorage;
        this.storage_auth_result = 'authorizationResult';
        this.storage_access_token = 'authorizationData';
        this.storage_id_token = 'authorizationDataIdToken';
        this.storage_is_authorized = '_isAuthorized';
        this.storage_user_data = 'userData';
        this.storage_auth_nonce = 'authNonce';
        this.storage_auth_state_control = 'authStateControl';
        this.storage_session_state = 'session_state';
        this.storage_silent_renew_running = 'storage_silent_renew_running';
        this.storage_custom_request_params = 'storage_custom_request_params';
    }
    /**
     * @return {?}
     */
    get authResult() {
        return this.retrieve(this.storage_auth_result);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set authResult(value) {
        this.store(this.storage_auth_result, value);
    }
    /**
     * @return {?}
     */
    get accessToken() {
        return this.retrieve(this.storage_access_token) || '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set accessToken(value) {
        this.store(this.storage_access_token, value);
    }
    /**
     * @return {?}
     */
    get idToken() {
        return this.retrieve(this.storage_id_token) || '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set idToken(value) {
        this.store(this.storage_id_token, value);
    }
    /**
     * @return {?}
     */
    get isAuthorized() {
        return this.retrieve(this.storage_is_authorized);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isAuthorized(value) {
        this.store(this.storage_is_authorized, value);
    }
    /**
     * @return {?}
     */
    get userData() {
        return this.retrieve(this.storage_user_data);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set userData(value) {
        this.store(this.storage_user_data, value);
    }
    /**
     * @return {?}
     */
    get authNonce() {
        return this.retrieve(this.storage_auth_nonce) || '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set authNonce(value) {
        this.store(this.storage_auth_nonce, value);
    }
    /**
     * @return {?}
     */
    get authStateControl() {
        return this.retrieve(this.storage_auth_state_control) || '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set authStateControl(value) {
        this.store(this.storage_auth_state_control, value);
    }
    /**
     * @return {?}
     */
    get sessionState() {
        return this.retrieve(this.storage_session_state);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set sessionState(value) {
        this.store(this.storage_session_state, value);
    }
    /**
     * @return {?}
     */
    get silentRenewRunning() {
        return this.retrieve(this.storage_silent_renew_running) || '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set silentRenewRunning(value) {
        this.store(this.storage_silent_renew_running, value);
    }
    /**
     * @return {?}
     */
    get customRequestParams() {
        return this.retrieve(this.storage_custom_request_params);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set customRequestParams(value) {
        this.store(this.storage_custom_request_params, value);
    }
    /**
     * @return {?}
     */
    setupModule() { }
    /**
     * @param {?} key
     * @return {?}
     */
    retrieve(key) {
        return this.oidcSecurityStorage.read(key);
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    store(key, value) {
        this.oidcSecurityStorage.write(key, value);
    }
    /**
     * @param {?} isRenewProcess
     * @return {?}
     */
    resetStorageData(isRenewProcess) {
        if (!isRenewProcess) {
            this.store(this.storage_auth_result, '');
            this.store(this.storage_session_state, '');
            this.store(this.storage_silent_renew_running, '');
            this.store(this.storage_is_authorized, false);
            this.store(this.storage_access_token, '');
            this.store(this.storage_id_token, '');
            this.store(this.storage_user_data, '');
        }
    }
    /**
     * @return {?}
     */
    getAccessToken() {
        return this.retrieve(this.storage_access_token);
    }
    /**
     * @return {?}
     */
    getIdToken() {
        return this.retrieve(this.storage_id_token);
    }
}
OidcSecurityCommon.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityCommon.ctorParameters = () => [
    { type: OidcSecurityStorage, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class ArrayHelperService {
    constructor() { }
    /**
     * @param {?} arr1
     * @param {?} arr2
     * @return {?}
     */
    arraysEqual(arr1, arr2) {
        if (arr1.length !== arr2.length) {
            return false;
        }
        for (let /** @type {?} */ i = arr1.length; i--;) {
            if (arr1[i] !== arr2[i]) {
                return false;
            }
        }
        return true;
    }
}
ArrayHelperService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
ArrayHelperService.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class TokenHelperService {
    constructor() { }
    /**
     * @param {?} dataIdToken
     * @return {?}
     */
    getTokenExpirationDate(dataIdToken) {
        if (!dataIdToken.hasOwnProperty('exp')) {
            return new Date();
        }
        const /** @type {?} */ date = new Date(0); // The 0 here is the key, which sets the date to the epoch
        date.setUTCSeconds(dataIdToken.exp);
        return date;
    }
    /**
     * @param {?} token
     * @param {?} encode
     * @return {?}
     */
    getPayloadFromToken(token, encode) {
        let /** @type {?} */ data = {};
        if (typeof token !== 'undefined') {
            const /** @type {?} */ encoded = token.split('.')[1];
            if (encode) {
                return encoded;
            }
            data = JSON.parse(this.urlBase64Decode(encoded));
        }
        return data;
    }
    /**
     * @param {?} token
     * @param {?} encode
     * @return {?}
     */
    getHeaderFromToken(token, encode) {
        let /** @type {?} */ data = {};
        if (typeof token !== 'undefined') {
            const /** @type {?} */ encoded = token.split('.')[0];
            if (encode) {
                return encoded;
            }
            data = JSON.parse(this.urlBase64Decode(encoded));
        }
        return data;
    }
    /**
     * @param {?} token
     * @param {?} encode
     * @return {?}
     */
    getSignatureFromToken(token, encode) {
        let /** @type {?} */ data = {};
        if (typeof token !== 'undefined') {
            const /** @type {?} */ encoded = token.split('.')[2];
            if (encode) {
                return encoded;
            }
            data = JSON.parse(this.urlBase64Decode(encoded));
        }
        return data;
    }
    /**
     * @param {?} str
     * @return {?}
     */
    urlBase64Decode(str) {
        let /** @type {?} */ output = str.replace('-', '+').replace('_', '/');
        switch (output.length % 4) {
            case 0:
                break;
            case 2:
                output += '==';
                break;
            case 3:
                output += '=';
                break;
            default:
                throw Error('Illegal base64url string!');
        }
        return window.atob(output);
    }
}
TokenHelperService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
TokenHelperService.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class LoggerService {
    /**
     * @param {?} authConfiguration
     */
    constructor(authConfiguration) {
        this.authConfiguration = authConfiguration;
    }
    /**
     * @param {?} message
     * @return {?}
     */
    logError(message) {
        console.error(message);
    }
    /**
     * @param {?} message
     * @return {?}
     */
    logWarning(message) {
        if (this.authConfiguration.log_console_warning_active) {
            console.warn(message);
        }
    }
    /**
     * @param {?} message
     * @return {?}
     */
    logDebug(message) {
        if (this.authConfiguration.log_console_debug_active) {
            console.log(message);
        }
    }
}
LoggerService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LoggerService.ctorParameters = () => [
    { type: AuthConfiguration, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcSecurityValidation {
    /**
     * @param {?} arrayHelperService
     * @param {?} tokenHelperService
     * @param {?} loggerService
     */
    constructor(arrayHelperService, tokenHelperService, loggerService) {
        this.arrayHelperService = arrayHelperService;
        this.tokenHelperService = tokenHelperService;
        this.loggerService = loggerService;
    }
    /**
     * @param {?} token
     * @param {?=} offsetSeconds
     * @return {?}
     */
    isTokenExpired(token, offsetSeconds) {
        let /** @type {?} */ decoded;
        decoded = this.tokenHelperService.getPayloadFromToken(token, false);
        return !this.validate_id_token_exp_not_expired(decoded, offsetSeconds);
    }
    /**
     * @param {?} decoded_id_token
     * @param {?=} offsetSeconds
     * @return {?}
     */
    validate_id_token_exp_not_expired(decoded_id_token, offsetSeconds) {
        const /** @type {?} */ tokenExpirationDate = this.tokenHelperService.getTokenExpirationDate(decoded_id_token);
        offsetSeconds = offsetSeconds || 0;
        if (!tokenExpirationDate) {
            return false;
        }
        // Token not expired?
        return (tokenExpirationDate.valueOf() >
            new Date().valueOf() + offsetSeconds * 1000);
    }
    /**
     * @param {?} dataIdToken
     * @return {?}
     */
    validate_required_id_token(dataIdToken) {
        let /** @type {?} */ validated = true;
        if (!dataIdToken.hasOwnProperty('iss')) {
            validated = false;
            this.loggerService.logWarning('iss is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('sub')) {
            validated = false;
            this.loggerService.logWarning('sub is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('aud')) {
            validated = false;
            this.loggerService.logWarning('aud is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('exp')) {
            validated = false;
            this.loggerService.logWarning('exp is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('iat')) {
            validated = false;
            this.loggerService.logWarning('iat is missing, this is required in the id_token');
        }
        return validated;
    }
    /**
     * @param {?} dataIdToken
     * @param {?} max_offset_allowed_in_seconds
     * @return {?}
     */
    validate_id_token_iat_max_offset(dataIdToken, max_offset_allowed_in_seconds) {
        if (!dataIdToken.hasOwnProperty('iat')) {
            return false;
        }
        const /** @type {?} */ dateTime_iat_id_token = new Date(0); // The 0 here is the key, which sets the date to the epoch
        dateTime_iat_id_token.setUTCSeconds(dataIdToken.iat);
        max_offset_allowed_in_seconds = max_offset_allowed_in_seconds || 0;
        if (dateTime_iat_id_token == null) {
            return false;
        }
        this.loggerService.logDebug('validate_id_token_iat_max_offset: ' +
            (new Date().valueOf() - dateTime_iat_id_token.valueOf()) +
            ' < ' +
            max_offset_allowed_in_seconds * 1000);
        return (new Date().valueOf() - dateTime_iat_id_token.valueOf() <
            max_offset_allowed_in_seconds * 1000);
    }
    /**
     * @param {?} dataIdToken
     * @param {?} local_nonce
     * @return {?}
     */
    validate_id_token_nonce(dataIdToken, local_nonce) {
        if (dataIdToken.nonce !== local_nonce) {
            this.loggerService.logDebug('Validate_id_token_nonce failed, dataIdToken.nonce: ' +
                dataIdToken.nonce +
                ' local_nonce:' +
                local_nonce);
            return false;
        }
        return true;
    }
    /**
     * @param {?} dataIdToken
     * @param {?} authWellKnownEndpoints_issuer
     * @return {?}
     */
    validate_id_token_iss(dataIdToken, authWellKnownEndpoints_issuer) {
        if ((/** @type {?} */ (dataIdToken.iss)) !==
            (/** @type {?} */ (authWellKnownEndpoints_issuer))) {
            this.loggerService.logDebug('Validate_id_token_iss failed, dataIdToken.iss: ' +
                dataIdToken.iss +
                ' authWellKnownEndpoints issuer:' +
                authWellKnownEndpoints_issuer);
            return false;
        }
        return true;
    }
    /**
     * @param {?} dataIdToken
     * @param {?} aud
     * @return {?}
     */
    validate_id_token_aud(dataIdToken, aud) {
        if (dataIdToken.aud instanceof Array) {
            const /** @type {?} */ result = this.arrayHelperService.arraysEqual(dataIdToken.aud, aud);
            if (!result) {
                this.loggerService.logDebug('Validate_id_token_aud  array failed, dataIdToken.aud: ' +
                    dataIdToken.aud +
                    ' client_id:' +
                    aud);
                return false;
            }
            return true;
        }
        else if (dataIdToken.aud !== aud) {
            this.loggerService.logDebug('Validate_id_token_aud failed, dataIdToken.aud: ' +
                dataIdToken.aud +
                ' client_id:' +
                aud);
            return false;
        }
        return true;
    }
    /**
     * @param {?} state
     * @param {?} local_state
     * @return {?}
     */
    validateStateFromHashCallback(state, local_state) {
        if ((/** @type {?} */ (state)) !== (/** @type {?} */ (local_state))) {
            this.loggerService.logDebug('ValidateStateFromHashCallback failed, state: ' +
                state +
                ' local_state:' +
                local_state);
            return false;
        }
        return true;
    }
    /**
     * @param {?} id_token_sub
     * @param {?} userdata_sub
     * @return {?}
     */
    validate_userdata_sub_id_token(id_token_sub, userdata_sub) {
        if ((/** @type {?} */ (id_token_sub)) !== (/** @type {?} */ (userdata_sub))) {
            this.loggerService.logDebug('validate_userdata_sub_id_token failed, id_token_sub: ' +
                id_token_sub +
                ' userdata_sub:' +
                userdata_sub);
            return false;
        }
        return true;
    }
    /**
     * @param {?} id_token
     * @param {?} jwtkeys
     * @return {?}
     */
    validate_signature_id_token(id_token, jwtkeys) {
        if (!jwtkeys || !jwtkeys.keys) {
            return false;
        }
        const /** @type {?} */ header_data = this.tokenHelperService.getHeaderFromToken(id_token, false);
        if (Object.keys(header_data).length === 0 &&
            header_data.constructor === Object) {
            this.loggerService.logWarning('id token has no header data');
            return false;
        }
        const /** @type {?} */ kid = header_data.kid;
        const /** @type {?} */ alg = header_data.alg;
        if ('RS256' !== (/** @type {?} */ (alg))) {
            this.loggerService.logWarning('Only RS256 supported');
            return false;
        }
        let /** @type {?} */ isValid = false;
        if (!header_data.hasOwnProperty('kid')) {
            // exactly 1 key in the jwtkeys and no kid in the Jose header
            // kty	"RSA" use "sig"
            let /** @type {?} */ amountOfMatchingKeys = 0;
            for (const /** @type {?} */ key of jwtkeys.keys) {
                if ((/** @type {?} */ (key.kty)) === 'RSA' &&
                    (/** @type {?} */ (key.use)) === 'sig') {
                    amountOfMatchingKeys = amountOfMatchingKeys + 1;
                }
            }
            if (amountOfMatchingKeys === 0) {
                this.loggerService.logWarning('no keys found, incorrect Signature, validation failed for id_token');
                return false;
            }
            else if (amountOfMatchingKeys > 1) {
                this.loggerService.logWarning('no ID Token kid claim in JOSE header and multiple supplied in jwks_uri');
                return false;
            }
            else {
                for (const /** @type {?} */ key of jwtkeys.keys) {
                    if ((/** @type {?} */ (key.kty)) === 'RSA' &&
                        (/** @type {?} */ (key.use)) === 'sig') {
                        const /** @type {?} */ publickey = KEYUTIL.getKey(key);
                        isValid = KJUR.jws.JWS.verify(id_token, publickey, [
                            'RS256'
                        ]);
                        if (!isValid) {
                            this.loggerService.logWarning('incorrect Signature, validation failed for id_token');
                        }
                        return isValid;
                    }
                }
            }
        }
        else {
            // kid in the Jose header of id_token
            for (const /** @type {?} */ key of jwtkeys.keys) {
                if ((/** @type {?} */ (key.kid)) === (/** @type {?} */ (kid))) {
                    const /** @type {?} */ publickey = KEYUTIL.getKey(key);
                    isValid = KJUR.jws.JWS.verify(id_token, publickey, [
                        'RS256'
                    ]);
                    if (!isValid) {
                        this.loggerService.logWarning('incorrect Signature, validation failed for id_token');
                    }
                    return isValid;
                }
            }
        }
        return isValid;
    }
    /**
     * @param {?} response_type
     * @return {?}
     */
    config_validate_response_type(response_type) {
        if (response_type === 'id_token token' ||
            response_type === 'id_token') {
            return true;
        }
        this.loggerService.logWarning('module configure incorrect, invalid response_type:' + response_type);
        return false;
    }
    /**
     * @param {?} access_token
     * @param {?} at_hash
     * @return {?}
     */
    validate_id_token_at_hash(access_token, at_hash) {
        this.loggerService.logDebug('From the server:' + at_hash);
        const /** @type {?} */ testdata = this.generate_at_hash('' + access_token);
        this.loggerService.logDebug('client validation not decoded:' + testdata);
        if (testdata === (/** @type {?} */ (at_hash))) {
            return true; // isValid;
        }
        else {
            const /** @type {?} */ testValue = this.generate_at_hash('' + decodeURIComponent(access_token));
            this.loggerService.logDebug('-gen access--' + testValue);
            if (testValue === (/** @type {?} */ (at_hash))) {
                return true; // isValid
            }
        }
        return false;
    }
    /**
     * @param {?} access_token
     * @return {?}
     */
    generate_at_hash(access_token) {
        const /** @type {?} */ hash = KJUR.crypto.Util.hashString(access_token, 'sha256');
        const /** @type {?} */ first128bits = hash.substr(0, hash.length / 2);
        const /** @type {?} */ testdata = hextob64u(first128bits);
        return testdata;
    }
}
OidcSecurityValidation.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityValidation.ctorParameters = () => [
    { type: ArrayHelperService, },
    { type: TokenHelperService, },
    { type: LoggerService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class StateValidationService {
    /**
     * @param {?} authConfiguration
     * @param {?} oidcSecurityCommon
     * @param {?} oidcSecurityValidation
     * @param {?} tokenHelperService
     * @param {?} loggerService
     */
    constructor(authConfiguration, oidcSecurityCommon, oidcSecurityValidation, tokenHelperService, loggerService) {
        this.authConfiguration = authConfiguration;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.oidcSecurityValidation = oidcSecurityValidation;
        this.tokenHelperService = tokenHelperService;
        this.loggerService = loggerService;
    }
    /**
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    setupModule(authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
    }
    /**
     * @param {?} result
     * @param {?} jwtKeys
     * @return {?}
     */
    validateState(result, jwtKeys) {
        const /** @type {?} */ toReturn = new ValidateStateResult('', '', false, {});
        if (!this.oidcSecurityValidation.validateStateFromHashCallback(result.state, this.oidcSecurityCommon.authStateControl)) {
            this.loggerService.logWarning('authorizedCallback incorrect state');
            return toReturn;
        }
        if (this.authConfiguration.response_type === 'id_token token') {
            toReturn.access_token = result.access_token;
        }
        toReturn.id_token = result.id_token;
        toReturn.decoded_id_token = this.tokenHelperService.getPayloadFromToken(toReturn.id_token, false);
        if (!this.oidcSecurityValidation.validate_signature_id_token(toReturn.id_token, jwtKeys)) {
            this.loggerService.logDebug('authorizedCallback Signature validation failed id_token');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_nonce(toReturn.decoded_id_token, this.oidcSecurityCommon.authNonce)) {
            this.loggerService.logWarning('authorizedCallback incorrect nonce');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_required_id_token(toReturn.decoded_id_token)) {
            this.loggerService.logDebug('authorizedCallback Validation, one of the REQUIRED properties missing from id_token');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_iat_max_offset(toReturn.decoded_id_token, this.authConfiguration
            .max_id_token_iat_offset_allowed_in_seconds)) {
            this.loggerService.logWarning('authorizedCallback Validation, iat rejected id_token was issued too far away from the current time');
            return toReturn;
        }
        if (this.authWellKnownEndpoints) {
            if (!this.oidcSecurityValidation.validate_id_token_iss(toReturn.decoded_id_token, this.authWellKnownEndpoints.issuer)) {
                this.loggerService.logWarning('authorizedCallback incorrect iss does not match authWellKnownEndpoints issuer');
                return toReturn;
            }
        }
        else {
            this.loggerService.logWarning('authWellKnownEndpoints is undefined');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_aud(toReturn.decoded_id_token, this.authConfiguration.client_id)) {
            this.loggerService.logWarning('authorizedCallback incorrect aud');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_exp_not_expired(toReturn.decoded_id_token)) {
            this.loggerService.logWarning('authorizedCallback token expired');
            return toReturn;
        }
        // flow id_token token
        if (this.authConfiguration.response_type !== 'id_token token') {
            toReturn.authResponseIsValid = true;
            this.handleSuccessfulValidation();
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_at_hash(toReturn.access_token, toReturn.decoded_id_token.at_hash) ||
            !toReturn.access_token) {
            this.loggerService.logWarning('authorizedCallback incorrect at_hash');
            return toReturn;
        }
        toReturn.authResponseIsValid = true;
        this.handleSuccessfulValidation();
        return toReturn;
    }
    /**
     * @return {?}
     */
    handleSuccessfulValidation() {
        this.oidcSecurityCommon.authNonce = '';
        if (this.authConfiguration.auto_clean_state_after_authentication) {
            this.oidcSecurityCommon.authStateControl = '';
        }
        this.loggerService.logDebug('AuthorizedCallback token(s) validated, continue');
    }
}
StateValidationService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
StateValidationService.ctorParameters = () => [
    { type: AuthConfiguration, },
    { type: OidcSecurityCommon, },
    { type: OidcSecurityValidation, },
    { type: TokenHelperService, },
    { type: LoggerService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcSecurityCheckSession {
    /**
     * @param {?} authConfiguration
     * @param {?} oidcSecurityCommon
     * @param {?} loggerService
     * @param {?} zone
     */
    constructor(authConfiguration, oidcSecurityCommon, loggerService, zone) {
        this.authConfiguration = authConfiguration;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.loggerService = loggerService;
        this.zone = zone;
        this.onCheckSessionChanged = new EventEmitter(true);
    }
    /**
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    setupModule(authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
    }
    /**
     * @return {?}
     */
    doesSessionExist() {
        let /** @type {?} */ existsparent = undefined;
        try {
            const /** @type {?} */ parentdoc = window.parent.document;
            if (!parentdoc) {
                throw new Error('Unaccessible');
            }
            existsparent = parentdoc.getElementById('myiFrameForCheckSession');
        }
        catch (/** @type {?} */ e) {
            // not accessible
        }
        const /** @type {?} */ exists = window.document.getElementById('myiFrameForCheckSession');
        if (existsparent) {
            this.sessionIframe = existsparent;
        }
        else if (exists) {
            this.sessionIframe = exists;
        }
        if (existsparent || exists) {
            return true;
        }
        return false;
    }
    /**
     * @return {?}
     */
    init() {
        this.sessionIframe = window.document.createElement('iframe');
        this.sessionIframe.id = 'myiFrameForCheckSession';
        this.loggerService.logDebug(this.sessionIframe);
        this.sessionIframe.style.display = 'none';
        window.document.body.appendChild(this.sessionIframe);
        if (this.authWellKnownEndpoints) {
            this.sessionIframe.src = this.authWellKnownEndpoints.check_session_iframe;
        }
        else {
            this.loggerService.logWarning('init check session: authWellKnownEndpoints is undefined');
        }
        this.iframeMessageEvent = this.messageHandler.bind(this);
        window.addEventListener('message', this.iframeMessageEvent, false);
        return Observable.create((observer) => {
            this.sessionIframe.onload = () => {
                observer.next(this);
                observer.complete();
            };
        });
    }
    /**
     * @param {?} clientId
     * @return {?}
     */
    startCheckingSession(clientId) {
        if (!this._scheduledHeartBeat) {
            this.pollServerSession(clientId);
        }
    }
    /**
     * @return {?}
     */
    stopCheckingSession() {
        if (this._scheduledHeartBeat) {
            clearTimeout(this._scheduledHeartBeat);
            this._scheduledHeartBeat = null;
        }
    }
    /**
     * @param {?} clientId
     * @return {?}
     */
    pollServerSession(clientId) {
        const /** @type {?} */ _pollServerSessionRecur = () => {
            if (this.sessionIframe && clientId) {
                this.loggerService.logDebug(this.sessionIframe);
                const /** @type {?} */ session_state = this.oidcSecurityCommon.sessionState;
                if (session_state) {
                    this.sessionIframe.contentWindow.postMessage(clientId + ' ' + session_state, this.authConfiguration.stsServer);
                }
            }
            else {
                this.loggerService.logWarning('OidcSecurityCheckSession pollServerSession sessionIframe does not exist');
                this.loggerService.logDebug(clientId);
                this.loggerService.logDebug(this.sessionIframe);
                // this.init();
            }
            this._scheduledHeartBeat = setTimeout(_pollServerSessionRecur, 3000);
        };
        this.zone.runOutsideAngular(() => {
            this._scheduledHeartBeat = setTimeout(_pollServerSessionRecur, 3000);
        });
    }
    /**
     * @param {?} e
     * @return {?}
     */
    messageHandler(e) {
        if (this.sessionIframe &&
            e.origin === this.authConfiguration.stsServer &&
            e.source === this.sessionIframe.contentWindow) {
            if (e.data === 'error') {
                this.loggerService.logWarning('error from checksession messageHandler');
            }
            else if (e.data === 'changed') {
                this.onCheckSessionChanged.emit();
            }
            else {
                this.loggerService.logDebug(e.data + ' from checksession messageHandler');
            }
        }
    }
}
OidcSecurityCheckSession.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityCheckSession.ctorParameters = () => [
    { type: AuthConfiguration, },
    { type: OidcSecurityCommon, },
    { type: LoggerService, },
    { type: NgZone, },
];
OidcSecurityCheckSession.propDecorators = {
    "onCheckSessionChanged": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcSecuritySilentRenew {
    /**
     * @param {?} loggerService
     */
    constructor(loggerService) {
        this.loggerService = loggerService;
    }
    /**
     * @return {?}
     */
    initRenew() {
        let /** @type {?} */ existsparent = undefined;
        try {
            const /** @type {?} */ parentdoc = window.parent.document;
            if (!parentdoc) {
                throw new Error('Unaccessible');
            }
            existsparent = parentdoc.getElementById('myiFrameForSilentRenew');
        }
        catch (/** @type {?} */ e) {
            // not accessible
        }
        const /** @type {?} */ exists = window.document.getElementById('myiFrameForSilentRenew');
        if (existsparent) {
            this.sessionIframe = existsparent;
        }
        else if (exists) {
            this.sessionIframe = exists;
        }
        if (!exists && !existsparent) {
            this.sessionIframe = window.document.createElement('iframe');
            this.sessionIframe.id = 'myiFrameForSilentRenew';
            this.loggerService.logDebug(this.sessionIframe);
            this.sessionIframe.style.display = 'none';
            window.document.body.appendChild(this.sessionIframe);
        }
    }
    /**
     * @param {?} url
     * @return {?}
     */
    startRenew(url) {
        let /** @type {?} */ existsparent = undefined;
        try {
            const /** @type {?} */ parentdoc = window.parent.document;
            if (!parentdoc) {
                throw new Error('Unaccessible');
            }
            existsparent = parentdoc.getElementById('myiFrameForSilentRenew');
        }
        catch (/** @type {?} */ e) {
            // not accessible
        }
        const /** @type {?} */ exists = window.document.getElementById('myiFrameForSilentRenew');
        if (existsparent) {
            this.sessionIframe = existsparent;
        }
        else if (exists) {
            this.sessionIframe = exists;
        }
        this.loggerService.logDebug('startRenew for URL:' + url);
        this.sessionIframe.src = url;
        return Observable.create((observer) => {
            this.sessionIframe.onload = () => {
                observer.next(this);
                observer.complete();
            };
        });
    }
}
OidcSecuritySilentRenew.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecuritySilentRenew.ctorParameters = () => [
    { type: LoggerService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcDataService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * @template T
     * @param {?} url
     * @return {?}
     */
    getWellknownEndpoints(url) {
        let /** @type {?} */ headers = new HttpHeaders();
        headers = headers.set('Accept', 'application/json');
        return this.httpClient.get(url, {
            headers: headers
        });
    }
    /**
     * @template T
     * @param {?} url
     * @param {?} token
     * @return {?}
     */
    getIdentityUserData(url, token) {
        let /** @type {?} */ headers = new HttpHeaders();
        headers = headers.set('Accept', 'application/json');
        headers = headers.set('Authorization', 'Bearer ' + decodeURIComponent(token));
        return this.httpClient.get(url, {
            headers: headers
        });
    }
    /**
     * @template T
     * @param {?} url
     * @return {?}
     */
    get(url) {
        let /** @type {?} */ headers = new HttpHeaders();
        headers = headers.set('Accept', 'application/json');
        return this.httpClient.get(url, {
            headers: headers
        });
    }
}
OidcDataService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcDataService.ctorParameters = () => [
    { type: HttpClient, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcSecurityUserService {
    /**
     * @param {?} oidcDataService
     * @param {?} oidcSecurityCommon
     * @param {?} loggerService
     */
    constructor(oidcDataService, oidcSecurityCommon, loggerService) {
        this.oidcDataService = oidcDataService;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.loggerService = loggerService;
        this.userData = '';
    }
    /**
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    setupModule(authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
    }
    /**
     * @return {?}
     */
    initUserData() {
        return this.getIdentityUserData().pipe(map((data) => (this.userData = data)));
    }
    /**
     * @return {?}
     */
    getUserData() {
        if (!this.userData) {
            throw Error('UserData is not set!');
        }
        return this.userData;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setUserData(value) {
        this.userData = value;
    }
    /**
     * @return {?}
     */
    getIdentityUserData() {
        const /** @type {?} */ token = this.oidcSecurityCommon.getAccessToken();
        if (this.authWellKnownEndpoints) {
            return this.oidcDataService.getIdentityUserData(this.authWellKnownEndpoints.userinfo_endpoint, token);
        }
        else {
            this.loggerService.logWarning('init check session: authWellKnownEndpoints is undefined');
        }
        return this.oidcDataService.getIdentityUserData('undefined', token);
    }
}
OidcSecurityUserService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityUserService.ctorParameters = () => [
    { type: OidcDataService, },
    { type: OidcSecurityCommon, },
    { type: LoggerService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class UriEncoder {
    /**
     * @param {?} key
     * @return {?}
     */
    encodeKey(key) {
        return encodeURIComponent(key);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    encodeValue(value) {
        return encodeURIComponent(value);
    }
    /**
     * @param {?} key
     * @return {?}
     */
    decodeKey(key) {
        return decodeURIComponent(key);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    decodeValue(value) {
        return decodeURIComponent(value);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class OidcSecurityService {
    /**
     * @param {?} platformId
     * @param {?} oidcDataService
     * @param {?} stateValidationService
     * @param {?} authConfiguration
     * @param {?} router
     * @param {?} oidcSecurityCheckSession
     * @param {?} oidcSecuritySilentRenew
     * @param {?} oidcSecurityUserService
     * @param {?} oidcSecurityCommon
     * @param {?} oidcSecurityValidation
     * @param {?} tokenHelperService
     * @param {?} loggerService
     * @param {?} zone
     */
    constructor(platformId, oidcDataService, stateValidationService, authConfiguration, router, oidcSecurityCheckSession, oidcSecuritySilentRenew, oidcSecurityUserService, oidcSecurityCommon, oidcSecurityValidation, tokenHelperService, loggerService, zone) {
        this.platformId = platformId;
        this.oidcDataService = oidcDataService;
        this.stateValidationService = stateValidationService;
        this.authConfiguration = authConfiguration;
        this.router = router;
        this.oidcSecurityCheckSession = oidcSecurityCheckSession;
        this.oidcSecuritySilentRenew = oidcSecuritySilentRenew;
        this.oidcSecurityUserService = oidcSecurityUserService;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.oidcSecurityValidation = oidcSecurityValidation;
        this.tokenHelperService = tokenHelperService;
        this.loggerService = loggerService;
        this.zone = zone;
        this.onModuleSetup = new EventEmitter();
        this.onAuthorizationResult = new EventEmitter();
        this.onCheckSessionChanged = new EventEmitter();
        this.checkSessionChanged = false;
        this.moduleSetup = false;
        this._isAuthorized = new BehaviorSubject(false);
        this._isAuthorizedValue = false;
        this._userData = new BehaviorSubject('');
        this.authWellKnownEndpointsLoaded = false;
        this.runTokenValidationRunning = false;
    }
    /**
     * @param {?} openIDImplicitFlowConfiguration
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    setupModule(openIDImplicitFlowConfiguration, authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
        this.authConfiguration.init(openIDImplicitFlowConfiguration);
        this.stateValidationService.setupModule(authWellKnownEndpoints);
        this.oidcSecurityCheckSession.setupModule(authWellKnownEndpoints);
        this.oidcSecurityUserService.setupModule(authWellKnownEndpoints);
        this.oidcSecurityCheckSession.onCheckSessionChanged.subscribe(() => {
            this.loggerService.logDebug('onCheckSessionChanged');
            this.checkSessionChanged = true;
            this.onCheckSessionChanged.emit(this.checkSessionChanged);
        });
        this._userData.subscribe(() => {
            this.onUserDataChanged();
        });
        const /** @type {?} */ userData = this.oidcSecurityCommon.userData;
        if (userData) {
            this.setUserData(userData);
        }
        const /** @type {?} */ isAuthorized = this.oidcSecurityCommon.isAuthorized;
        if (isAuthorized) {
            this.loggerService.logDebug('IsAuthorized setup module');
            this.loggerService.logDebug(this.oidcSecurityCommon.idToken);
            if (this.oidcSecurityValidation.isTokenExpired(this.oidcSecurityCommon.idToken, this.authConfiguration.silent_renew_offset_in_seconds)) {
                this.loggerService.logDebug('IsAuthorized setup module; id_token isTokenExpired');
            }
            else {
                this.loggerService.logDebug('IsAuthorized setup module; id_token is valid');
                this.setIsAuthorized(isAuthorized);
                this.runTokenValidation();
            }
        }
        this.loggerService.logDebug('STS server: ' + this.authConfiguration.stsServer);
        if (isPlatformBrowser(this.platformId)) {
            // Client only code.
            this.moduleSetup = true;
            this.onModuleSetup.emit();
            if (this.authConfiguration.silent_renew) {
                this.oidcSecuritySilentRenew.initRenew();
                // Support authorization via DOM events.
                // Deregister if OidcSecurityService.setupModule is called again by any instance.
                //      We only ever want the latest setup service to be reacting to this event.
                this.boundSilentRenewEvent = this.silentRenewEventHandler.bind(this);
                const /** @type {?} */ instanceId = Math.random();
                const /** @type {?} */ boundSilentRenewInitEvent = ((e) => {
                    if (e.detail !== instanceId) {
                        window.removeEventListener('oidc-silent-renew-message', this.boundSilentRenewEvent);
                        window.removeEventListener('oidc-silent-renew-init', boundSilentRenewInitEvent);
                    }
                }).bind(this);
                window.addEventListener('oidc-silent-renew-init', boundSilentRenewInitEvent, false);
                window.addEventListener('oidc-silent-renew-message', this.boundSilentRenewEvent, false);
                window.dispatchEvent(new CustomEvent('oidc-silent-renew-init', { detail: instanceId }));
            }
            if (this.authConfiguration.start_checksession &&
                !this.oidcSecurityCheckSession.doesSessionExist()) {
                this.oidcSecurityCheckSession.init().subscribe(() => {
                    this.oidcSecurityCheckSession.pollServerSession(this.authConfiguration.client_id);
                });
            }
        }
        else {
            this.moduleSetup = true;
            this.onModuleSetup.emit();
        }
    }
    /**
     * @return {?}
     */
    getUserData() {
        return this._userData.asObservable();
    }
    /**
     * @return {?}
     */
    getIsAuthorized() {
        return this._isAuthorized.asObservable();
    }
    /**
     * @return {?}
     */
    getToken() {
        if (!this._isAuthorizedValue) {
            return '';
        }
        const /** @type {?} */ token = this.oidcSecurityCommon.getAccessToken();
        return decodeURIComponent(token);
    }
    /**
     * @return {?}
     */
    getIdToken() {
        if (!this._isAuthorizedValue) {
            return '';
        }
        const /** @type {?} */ token = this.oidcSecurityCommon.getIdToken();
        return decodeURIComponent(token);
    }
    /**
     * @param {?=} encode
     * @return {?}
     */
    getPayloadFromIdToken(encode = false) {
        const /** @type {?} */ token = this.getIdToken();
        return this.tokenHelperService.getPayloadFromToken(token, encode);
    }
    /**
     * @param {?} state
     * @return {?}
     */
    setState(state) {
        this.oidcSecurityCommon.authStateControl = state;
    }
    /**
     * @return {?}
     */
    getState() {
        return this.oidcSecurityCommon.authStateControl;
    }
    /**
     * @param {?} params
     * @return {?}
     */
    setCustomRequestParameters(params) {
        this.oidcSecurityCommon.customRequestParams = params;
    }
    /**
     * @param {?=} urlHandler
     * @return {?}
     */
    authorize(urlHandler) {
        if (this.authWellKnownEndpoints) {
            this.authWellKnownEndpointsLoaded = true;
        }
        if (!this.authWellKnownEndpointsLoaded) {
            this.loggerService.logError('Well known endpoints must be loaded before user can login!');
            return;
        }
        if (!this.oidcSecurityValidation.config_validate_response_type(this.authConfiguration.response_type)) {
            // invalid response_type
            return;
        }
        this.resetAuthorizationData(false);
        this.loggerService.logDebug('BEGIN Authorize, no auth data');
        let /** @type {?} */ state = this.oidcSecurityCommon.authStateControl;
        if (!state) {
            state = Date.now() + '' + Math.random();
            this.oidcSecurityCommon.authStateControl = state;
        }
        const /** @type {?} */ nonce = 'N' + Math.random() + '' + Date.now();
        this.oidcSecurityCommon.authNonce = nonce;
        this.loggerService.logDebug('AuthorizedController created. local state: ' +
            this.oidcSecurityCommon.authStateControl);
        if (this.authWellKnownEndpoints) {
            const /** @type {?} */ url = this.createAuthorizeUrl(this.authConfiguration.redirect_url, nonce, state, this.authWellKnownEndpoints.authorization_endpoint);
            if (urlHandler) {
                urlHandler(url);
            }
            else {
                window.location.href = url;
            }
        }
        else {
            this.loggerService.logError('authWellKnownEndpoints is undefined');
        }
    }
    /**
     * @param {?=} hash
     * @return {?}
     */
    authorizedCallback(hash) {
        const /** @type {?} */ silentRenew = this.oidcSecurityCommon.silentRenewRunning;
        const /** @type {?} */ isRenewProcess = silentRenew === 'running';
        this.loggerService.logDebug('BEGIN authorizedCallback, no auth data');
        this.resetAuthorizationData(isRenewProcess);
        hash = hash || window.location.hash.substr(1);
        const /** @type {?} */ result = hash
            .split('&')
            .reduce(function (resultData, item) {
            const /** @type {?} */ parts = item.split('=');
            resultData[parts[0]] = parts[1];
            return resultData;
        }, {});
        this.oidcSecurityCommon.authResult = result;
        this.loggerService.logDebug(result);
        this.loggerService.logDebug('authorizedCallback created, begin token validation');
        this.getSigningKeys().subscribe(jwtKeys => {
            const /** @type {?} */ validationResult = this.getValidatedStateResult(result, jwtKeys);
            if (validationResult.authResponseIsValid) {
                this.setAuthorizationData(validationResult.access_token, validationResult.id_token);
                this.oidcSecurityCommon.silentRenewRunning = '';
                if (this.authConfiguration.auto_userinfo) {
                    this.getUserinfo(isRenewProcess, result, validationResult.id_token, validationResult.decoded_id_token).subscribe(response => {
                        if (response) {
                            this.onAuthorizationResult.emit(AuthorizationResult.authorized);
                            if (!this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                                this.router.navigate([
                                    this.authConfiguration.post_login_route
                                ]);
                            }
                        }
                        else {
                            this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
                            if (!this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                                this.router.navigate([
                                    this.authConfiguration.unauthorized_route
                                ]);
                            }
                        }
                    });
                }
                else {
                    if (!isRenewProcess) {
                        // userData is set to the id_token decoded, auto get user data set to false
                        this.oidcSecurityUserService.setUserData(validationResult.decoded_id_token);
                        this.setUserData(this.oidcSecurityUserService.getUserData());
                        this.runTokenValidation();
                    }
                    this.onAuthorizationResult.emit(AuthorizationResult.authorized);
                    if (!this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                        this.router.navigate([
                            this.authConfiguration.post_login_route
                        ]);
                    }
                }
            }
            else {
                // something went wrong
                this.loggerService.logWarning('authorizedCallback, token(s) validation failed, resetting');
                this.loggerService.logWarning(window.location.hash);
                this.resetAuthorizationData(false);
                this.oidcSecurityCommon.silentRenewRunning = '';
                this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
                if (!this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                    this.router.navigate([
                        this.authConfiguration.unauthorized_route
                    ]);
                }
            }
        }, (err) => {
            /* Something went wrong while getting signing key */
            this.loggerService.logWarning('Failed to retreive siging key with error: ' + JSON.stringify(err));
            this.oidcSecurityCommon.silentRenewRunning = '';
        });
    }
    /**
     * @param {?=} isRenewProcess
     * @param {?=} result
     * @param {?=} id_token
     * @param {?=} decoded_id_token
     * @return {?}
     */
    getUserinfo(isRenewProcess = false, result, id_token, decoded_id_token) {
        result = result ? result : this.oidcSecurityCommon.authResult;
        id_token = id_token ? id_token : this.oidcSecurityCommon.idToken;
        decoded_id_token = decoded_id_token
            ? decoded_id_token
            : this.tokenHelperService.getPayloadFromToken(id_token, false);
        return new Observable(observer => {
            // flow id_token token
            if (this.authConfiguration.response_type === 'id_token token') {
                if (isRenewProcess) {
                    this.oidcSecurityCommon.sessionState = result.session_state;
                    observer.next(true);
                    observer.complete();
                }
                else {
                    this.oidcSecurityUserService
                        .initUserData()
                        .subscribe(() => {
                        this.loggerService.logDebug('authorizedCallback id_token token flow');
                        const /** @type {?} */ userData = this.oidcSecurityUserService.getUserData();
                        if (this.oidcSecurityValidation.validate_userdata_sub_id_token(decoded_id_token.sub, userData.sub)) {
                            this.setUserData(userData);
                            this.loggerService.logDebug(this.oidcSecurityCommon.accessToken);
                            this.loggerService.logDebug(this.oidcSecurityUserService.getUserData());
                            this.oidcSecurityCommon.sessionState =
                                result.session_state;
                            this.runTokenValidation();
                            observer.next(true);
                        }
                        else {
                            // something went wrong, userdata sub does not match that from id_token
                            this.loggerService.logWarning('authorizedCallback, User data sub does not match sub in id_token');
                            this.loggerService.logDebug('authorizedCallback, token(s) validation failed, resetting');
                            this.resetAuthorizationData(false);
                            observer.next(false);
                        }
                        observer.complete();
                    });
                }
            }
            else {
                // flow id_token
                this.loggerService.logDebug('authorizedCallback id_token flow');
                this.loggerService.logDebug(this.oidcSecurityCommon.accessToken);
                // userData is set to the id_token decoded. No access_token.
                this.oidcSecurityUserService.setUserData(decoded_id_token);
                this.setUserData(this.oidcSecurityUserService.getUserData());
                this.oidcSecurityCommon.sessionState = result.session_state;
                if (!isRenewProcess) {
                    this.runTokenValidation();
                }
                observer.next(true);
                observer.complete();
            }
        });
    }
    /**
     * @return {?}
     */
    logoff() {
        // /connect/endsession?id_token_hint=...&post_logout_redirect_uri=https://myapp.com
        this.loggerService.logDebug('BEGIN Authorize, no auth data');
        if (this.authWellKnownEndpoints) {
            if (this.authWellKnownEndpoints.end_session_endpoint) {
                const /** @type {?} */ end_session_endpoint = this.authWellKnownEndpoints
                    .end_session_endpoint;
                const /** @type {?} */ id_token_hint = this.oidcSecurityCommon.idToken;
                const /** @type {?} */ url = this.createEndSessionUrl(end_session_endpoint, id_token_hint);
                this.resetAuthorizationData(false);
                if (this.authConfiguration.start_checksession &&
                    this.checkSessionChanged) {
                    this.loggerService.logDebug('only local login cleaned up, server session has changed');
                }
                else {
                    window.location.href = url;
                }
            }
            else {
                this.resetAuthorizationData(false);
                this.loggerService.logDebug('only local login cleaned up, no end_session_endpoint');
            }
        }
        else {
            this.loggerService.logWarning('authWellKnownEndpoints is undefined');
        }
    }
    /**
     * @return {?}
     */
    refreshSession() {
        this.loggerService.logDebug('BEGIN refresh session Authorize');
        let /** @type {?} */ state = this.oidcSecurityCommon.authStateControl;
        if (state === '' || state === null) {
            state = Date.now() + '' + Math.random();
            this.oidcSecurityCommon.authStateControl = state;
        }
        const /** @type {?} */ nonce = 'N' + Math.random() + '' + Date.now();
        this.oidcSecurityCommon.authNonce = nonce;
        this.loggerService.logDebug('RefreshSession created. adding myautostate: ' +
            this.oidcSecurityCommon.authStateControl);
        let /** @type {?} */ url = '';
        if (this.authWellKnownEndpoints) {
            url = this.createAuthorizeUrl(this.authConfiguration.silent_redirect_url, nonce, state, this.authWellKnownEndpoints.authorization_endpoint, 'none');
        }
        else {
            this.loggerService.logWarning('authWellKnownEndpoints is undefined');
        }
        this.oidcSecurityCommon.silentRenewRunning = 'running';
        return this.oidcSecuritySilentRenew.startRenew(url);
    }
    /**
     * @param {?} error
     * @return {?}
     */
    handleError(error) {
        this.loggerService.logError(error);
        if (error.status === 403 || error.status === '403') {
            if (this.authConfiguration.trigger_authorization_result_event) {
                this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
            }
            else {
                this.router.navigate([this.authConfiguration.forbidden_route]);
            }
        }
        else if (error.status === 401 || error.status === '401') {
            const /** @type {?} */ silentRenew = this.oidcSecurityCommon.silentRenewRunning;
            this.resetAuthorizationData(!!silentRenew);
            if (this.authConfiguration.trigger_authorization_result_event) {
                this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
            }
            else {
                this.router.navigate([
                    this.authConfiguration.unauthorized_route
                ]);
            }
        }
    }
    /**
     * @return {?}
     */
    startCheckingSilentRenew() {
        this.runTokenValidation();
    }
    /**
     * @return {?}
     */
    stopCheckingSilentRenew() {
        if (this._scheduledHeartBeat) {
            clearTimeout(this._scheduledHeartBeat);
            this._scheduledHeartBeat = null;
            this.runTokenValidationRunning = false;
        }
    }
    /**
     * @param {?} result
     * @param {?} jwtKeys
     * @return {?}
     */
    getValidatedStateResult(result, jwtKeys) {
        if (result.error) {
            return new ValidateStateResult('', '', false, {});
        }
        return this.stateValidationService.validateState(result, jwtKeys);
    }
    /**
     * @param {?} userData
     * @return {?}
     */
    setUserData(userData) {
        this.oidcSecurityCommon.userData = userData;
        this._userData.next(userData);
    }
    /**
     * @param {?} isAuthorized
     * @return {?}
     */
    setIsAuthorized(isAuthorized) {
        this._isAuthorizedValue = isAuthorized;
        this._isAuthorized.next(isAuthorized);
    }
    /**
     * @param {?} access_token
     * @param {?} id_token
     * @return {?}
     */
    setAuthorizationData(access_token, id_token) {
        if (this.oidcSecurityCommon.accessToken !== '') {
            this.oidcSecurityCommon.accessToken = '';
        }
        this.loggerService.logDebug(access_token);
        this.loggerService.logDebug(id_token);
        this.loggerService.logDebug('storing to storage, getting the roles');
        this.oidcSecurityCommon.accessToken = access_token;
        this.oidcSecurityCommon.idToken = id_token;
        this.setIsAuthorized(true);
        this.oidcSecurityCommon.isAuthorized = true;
    }
    /**
     * @param {?} redirect_url
     * @param {?} nonce
     * @param {?} state
     * @param {?} authorization_endpoint
     * @param {?=} prompt
     * @return {?}
     */
    createAuthorizeUrl(redirect_url, nonce, state, authorization_endpoint, prompt) {
        const /** @type {?} */ urlParts = authorization_endpoint.split('?');
        const /** @type {?} */ authorizationUrl = urlParts[0];
        let /** @type {?} */ params = new HttpParams({
            fromString: urlParts[1],
            encoder: new UriEncoder()
        });
        params = params.set('client_id', this.authConfiguration.client_id);
        params = params.append('redirect_uri', redirect_url);
        params = params.append('response_type', this.authConfiguration.response_type);
        params = params.append('scope', this.authConfiguration.scope);
        params = params.append('nonce', nonce);
        params = params.append('state', state);
        if (prompt) {
            params = params.append('prompt', prompt);
        }
        if (this.authConfiguration.hd_param) {
            params = params.append('hd', this.authConfiguration.hd_param);
        }
        const /** @type {?} */ customParams = Object.assign({}, this.oidcSecurityCommon.customRequestParams);
        Object.keys(customParams).forEach(key => {
            params = params.append(key, customParams[key].toString());
        });
        return `${authorizationUrl}?${params}`;
    }
    /**
     * @param {?} end_session_endpoint
     * @param {?} id_token_hint
     * @return {?}
     */
    createEndSessionUrl(end_session_endpoint, id_token_hint) {
        const /** @type {?} */ urlParts = end_session_endpoint.split('?');
        const /** @type {?} */ authorizationEndsessionUrl = urlParts[0];
        let /** @type {?} */ params = new HttpParams({
            fromString: urlParts[1],
            encoder: new UriEncoder()
        });
        params = params.set('id_token_hint', id_token_hint);
        params = params.append('post_logout_redirect_uri', this.authConfiguration.post_logout_redirect_uri);
        return `${authorizationEndsessionUrl}?${params}`;
    }
    /**
     * @param {?} isRenewProcess
     * @return {?}
     */
    resetAuthorizationData(isRenewProcess) {
        if (!isRenewProcess) {
            if (this.authConfiguration.auto_userinfo) {
                // Clear user data. Fixes #97.
                this.setUserData('');
            }
            this.setIsAuthorized(false);
            this.oidcSecurityCommon.resetStorageData(isRenewProcess);
            this.checkSessionChanged = false;
        }
    }
    /**
     * @return {?}
     */
    onUserDataChanged() {
        this.loggerService.logDebug(`onUserDataChanged: last = ${this.lastUserData}, new = ${this._userData.value}`);
        if (this.lastUserData && !this._userData.value) {
            this.loggerService.logDebug('onUserDataChanged: Logout detected.');
            // TODO should we have an action here
        }
        this.lastUserData = this._userData.value;
    }
    /**
     * @return {?}
     */
    getSigningKeys() {
        if (this.authWellKnownEndpoints) {
            this.loggerService.logDebug('jwks_uri: ' + this.authWellKnownEndpoints.jwks_uri);
            return this.oidcDataService
                .get(this.authWellKnownEndpoints.jwks_uri)
                .pipe(catchError(this.handleErrorGetSigningKeys));
        }
        else {
            this.loggerService.logWarning('getSigningKeys: authWellKnownEndpoints is undefined');
        }
        return this.oidcDataService
            .get('undefined')
            .pipe(catchError(this.handleErrorGetSigningKeys));
    }
    /**
     * @param {?} error
     * @return {?}
     */
    handleErrorGetSigningKeys(error) {
        let /** @type {?} */ errMsg;
        if (error instanceof Response) {
            const /** @type {?} */ body = error.json() || {};
            const /** @type {?} */ err = JSON.stringify(body);
            errMsg = `${error.status} - ${error.statusText || ''} ${err}`;
        }
        else {
            errMsg = error.message ? error.message : error.toString();
        }
        console.error(errMsg);
        return throwError(errMsg);
    }
    /**
     * @return {?}
     */
    runTokenValidation() {
        if (this.runTokenValidationRunning) {
            return;
        }
        this.runTokenValidationRunning = true;
        /**
         * First time: delay 10 seconds to call silentRenewHeartBeatCheck
         * Afterwards: Run this check in a 5 second interval only AFTER the previous operation ends.
         */
        const /** @type {?} */ silentRenewHeartBeatCheck = () => {
            if (this._userData.value && (this.oidcSecurityCommon.silentRenewRunning !== 'running') && this.getIdToken()) {
                if (this.oidcSecurityValidation.isTokenExpired(this.oidcSecurityCommon.idToken, this.authConfiguration.silent_renew_offset_in_seconds)) {
                    this.loggerService.logDebug('IsAuthorized: id_token isTokenExpired, start silent renew if active');
                    if (this.authConfiguration.silent_renew) {
                        this.refreshSession().subscribe(() => {
                            this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 3000);
                        }, (err) => {
                            this.loggerService.logError('Error: ' + err);
                            this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 3000);
                        });
                        /* In this situation, we schedule a heatbeat check only when silentRenew is finished.
                                                We don't want to schedule another check so we have to return here */
                        return;
                    }
                    else {
                        this.resetAuthorizationData(false);
                    }
                }
            }
            /* Delay 3 seconds and do the next check */
            this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 3000);
        };
        this.zone.runOutsideAngular(() => {
            /* Initial heartbeat check */
            this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 10000);
        });
    }
    /**
     * @param {?} e
     * @return {?}
     */
    silentRenewEventHandler(e) {
        this.loggerService.logDebug('silentRenewEventHandler');
        this.authorizedCallback(e.detail);
    }
}
OidcSecurityService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityService.ctorParameters = () => [
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
    { type: OidcDataService, },
    { type: StateValidationService, },
    { type: AuthConfiguration, },
    { type: Router, },
    { type: OidcSecurityCheckSession, },
    { type: OidcSecuritySilentRenew, },
    { type: OidcSecurityUserService, },
    { type: OidcSecurityCommon, },
    { type: OidcSecurityValidation, },
    { type: TokenHelperService, },
    { type: LoggerService, },
    { type: NgZone, },
];
OidcSecurityService.propDecorators = {
    "onModuleSetup": [{ type: Output },],
    "onAuthorizationResult": [{ type: Output },],
    "onCheckSessionChanged": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class AuthWellKnownEndpoints {
    constructor() {
        this.issuer = '';
        this.jwks_uri = '';
        this.authorization_endpoint = '';
        this.token_endpoint = '';
        this.userinfo_endpoint = '';
        this.end_session_endpoint = '';
        this.check_session_iframe = '';
        this.revocation_endpoint = '';
        this.introspection_endpoint = '';
    }
    /**
     * @param {?} data
     * @return {?}
     */
    setWellKnownEndpoints(data) {
        this.issuer = data.issuer;
        this.jwks_uri = data.jwks_uri;
        this.authorization_endpoint = data.authorization_endpoint;
        this.token_endpoint = data.token_endpoint;
        this.userinfo_endpoint = data.userinfo_endpoint;
        if (data.end_session_endpoint) {
            this.end_session_endpoint = data.end_session_endpoint;
        }
        if (data.check_session_iframe) {
            this.check_session_iframe = data.check_session_iframe;
        }
        if (data.revocation_endpoint) {
            this.revocation_endpoint = data.revocation_endpoint;
        }
        if (data.introspection_endpoint) {
            this.introspection_endpoint = data.introspection_endpoint;
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class JwtKeys {
    constructor() {
        this.keys = [];
    }
}
class JwtKey {
    constructor() {
        this.kty = '';
        this.use = '';
        this.kid = '';
        this.x5t = '';
        this.e = '';
        this.n = '';
        this.x5c = [];
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class AuthModule {
    /**
     * @param {?=} token
     * @return {?}
     */
    static forRoot(token = {}) {
        return {
            ngModule: AuthModule,
            providers: [
                OidcConfigService,
                OidcSecurityService,
                OidcSecurityValidation,
                OidcSecurityCheckSession,
                OidcSecuritySilentRenew,
                OidcSecurityUserService,
                OidcSecurityCommon,
                AuthConfiguration,
                TokenHelperService,
                LoggerService,
                DefaultConfiguration,
                ArrayHelperService,
                AuthWellKnownEndpoints,
                OidcDataService,
                StateValidationService,
                {
                    provide: OidcSecurityStorage,
                    useClass: token.storage || BrowserStorage
                }
            ]
        };
    }
}
AuthModule.decorators = [
    { type: NgModule },
];
/**
 * @record
 * @template T
 */

/**
 * @record
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// Public classes.

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Angular OpenID Connect Implicit Flow.
 * OpenID Connect Implicit Flow RP Client
 * Written by Damien Bowden.
 * MIT license.
 * https://github.com/damienbod/angular-auth-oidc-client
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { OidcConfigService, OidcSecurityService, OidcSecurityStorage, BrowserStorage, AuthWellKnownEndpoints, AuthorizationResult, JwtKeys, JwtKey, ValidateStateResult, DefaultConfiguration, OpenIDImplicitFlowConfiguration, AuthConfiguration, AuthModule, OidcSecurityValidation, ArrayHelperService as d, OidcDataService as a, StateValidationService as b, TokenHelperService as e, LoggerService as f, OidcSecurityCheckSession as g, OidcSecurityCommon as c, OidcSecuritySilentRenew as h, OidcSecurityUserService as i };
