// angular-auth-oidc-client/modules/angular-auth-oidc-client.es5.js
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { EventEmitter, Inject, Injectable, NgModule, NgZone, Output, PLATFORM_ID } from '@angular/core';
import { BehaviorSubject, Observable, throwError } from 'rxjs';
import { isPlatformBrowser } from '@angular/common';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Router } from '@angular/router';
import { catchError, map } from 'rxjs/operators';
import { KEYUTIL, KJUR, hextob64u } from 'jsrsasign';
var __awaiter = (undefined && undefined.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try {
            step(generator.next(value));
        }
        catch (e) {
            reject(e);
        } }
        function rejected(value) { try {
            step(generator["throw"](value));
        }
        catch (e) {
            reject(e);
        } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcConfigService = /** @class */ (function () {
    function OidcConfigService() {
        this.onConfigurationLoaded = new EventEmitter();
    }
    /**
     * @param {?} configUrl
     * @return {?}
     */
    OidcConfigService.prototype.load = function (configUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var response, _a, err_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 4, , 5]);
                        return [4 /*yield*/, fetch(configUrl)];
                    case 1:
                        response = _b.sent();
                        if (!response.ok) {
                            throw new Error(response.statusText);
                        }
                        _a = this;
                        return [4 /*yield*/, response.json()];
                    case 2:
                        _a.clientConfiguration = _b.sent();
                        return [4 /*yield*/, this.load_using_stsServer(this.clientConfiguration.stsServer)];
                    case 3:
                        _b.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        err_1 = _b.sent();
                        console.error("OidcConfigService 'load' threw an error on calling " + configUrl, err_1);
                        this.onConfigurationLoaded.emit(false);
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @param {?} stsServer
     * @return {?}
     */
    OidcConfigService.prototype.load_using_stsServer = function (stsServer) {
        return __awaiter(this, void 0, void 0, function () {
            var response, _a, err_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, fetch(stsServer + "/.well-known/openid-configuration")];
                    case 1:
                        response = _b.sent();
                        if (!response.ok) {
                            throw new Error(response.statusText);
                        }
                        _a = this;
                        return [4 /*yield*/, response.json()];
                    case 2:
                        _a.wellKnownEndpoints = _b.sent();
                        this.onConfigurationLoaded.emit(true);
                        return [3 /*break*/, 4];
                    case 3:
                        err_2 = _b.sent();
                        console.error("OidcConfigService 'load_using_stsServer' threw an error on calling " + stsServer, err_2);
                        this.onConfigurationLoaded.emit(false);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @param {?} stsServer
     * @return {?}
     */
    OidcConfigService.prototype.load_using_custom_stsServer = function (stsServer) {
        return __awaiter(this, void 0, void 0, function () {
            var response, _a, err_3;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, fetch(stsServer)];
                    case 1:
                        response = _b.sent();
                        if (!response.ok) {
                            throw new Error(response.statusText);
                        }
                        _a = this;
                        return [4 /*yield*/, response.json()];
                    case 2:
                        _a.wellKnownEndpoints = _b.sent();
                        this.onConfigurationLoaded.emit(true);
                        return [3 /*break*/, 4];
                    case 3:
                        err_3 = _b.sent();
                        console.error("OidcConfigService 'load_using_custom_stsServer' threw an error on calling " + stsServer, err_3);
                        this.onConfigurationLoaded.emit(false);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return OidcConfigService;
}());
OidcConfigService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcConfigService.propDecorators = {
    "onConfigurationLoaded": [{ type: Output },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {number} */
var AuthorizationResult = {
    authorized: 1,
    forbidden: 2,
    unauthorized: 3,
};
AuthorizationResult[AuthorizationResult.authorized] = "authorized";
AuthorizationResult[AuthorizationResult.forbidden] = "forbidden";
AuthorizationResult[AuthorizationResult.unauthorized] = "unauthorized";
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var ValidateStateResult = /** @class */ (function () {
    /**
     * @param {?=} access_token
     * @param {?=} id_token
     * @param {?=} authResponseIsValid
     * @param {?=} decoded_id_token
     */
    function ValidateStateResult(access_token, id_token, authResponseIsValid, decoded_id_token) {
        if (access_token === void 0) { access_token = ''; }
        if (id_token === void 0) { id_token = ''; }
        if (authResponseIsValid === void 0) { authResponseIsValid = false; }
        this.access_token = access_token;
        this.id_token = id_token;
        this.authResponseIsValid = authResponseIsValid;
        this.decoded_id_token = decoded_id_token;
    }
    return ValidateStateResult;
}());
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var DefaultConfiguration = /** @class */ (function () {
    function DefaultConfiguration() {
        this.stsServer = 'https://localhost:44318';
        this.redirect_url = 'https://localhost:44311';
        // The Client MUST validate that the aud (audience) Claim contains its client_id value registered at the Issuer identified
        // by the iss (issuer) Claim as an audience.
        // The ID Token MUST be rejected if the ID Token does not list the Client as a valid audience,
        // or if it contains additional audiences not trusted by the Client.
        this.client_id = 'angularclient';
        this.response_type = 'id_token token';
        this.scope = 'openid email profile';
        // Only for Google Auth with particular G Suite domain, see https://developers.google.com/identity/protocols/OpenIDConnect#hd-param
        this.hd_param = '';
        this.post_logout_redirect_uri = 'https://localhost:44311/unauthorized';
        this.start_checksession = false;
        this.silent_renew = true;
        this.silent_renew_url = 'https://localhost:44311';
        this.silent_renew_offset_in_seconds = 0;
        this.silent_redirect_url = 'https://localhost:44311';
        this.post_login_route = '/';
        // HTTP 403
        this.forbidden_route = '/forbidden';
        // HTTP 401
        this.unauthorized_route = '/unauthorized';
        this.auto_userinfo = true;
        this.auto_clean_state_after_authentication = true;
        this.trigger_authorization_result_event = false;
        this.log_console_warning_active = true;
        this.log_console_debug_active = false;
        // id_token C8: The iat Claim can be used to reject tokens that were issued too far away from the current time,
        // limiting the amount of time that nonces need to be stored to prevent attacks.The acceptable range is Client specific.
        this.max_id_token_iat_offset_allowed_in_seconds = 3;
        this.storage = typeof Storage !== 'undefined' ? sessionStorage : null;
    }
    return DefaultConfiguration;
}());
var OpenIDImplicitFlowConfiguration = /** @class */ (function () {
    function OpenIDImplicitFlowConfiguration() {
        this.stsServer = 'https://localhost:44318';
        this.redirect_url = 'https://localhost:44311';
        this.client_id = 'angularclient';
        this.response_type = 'id_token token';
        this.resource = '';
        this.scope = 'openid email profile';
        this.hd_param = '';
        this.post_logout_redirect_uri = 'https://localhost:44311/unauthorized';
        this.start_checksession = false;
        this.silent_renew = true;
        this.silent_renew_url = 'https://localhost:44311';
        this.silent_renew_offset_in_seconds = 0;
        this.silent_redirect_url = 'https://localhost:44311';
        this.post_login_route = '/';
        this.forbidden_route = '/forbidden';
        this.unauthorized_route = '/unauthorized';
        this.auto_userinfo = true;
        this.auto_clean_state_after_authentication = true;
        this.trigger_authorization_result_event = false;
        this.log_console_warning_active = true;
        this.log_console_debug_active = false;
        this.max_id_token_iat_offset_allowed_in_seconds = 3;
        this.storage = sessionStorage;
    }
    return OpenIDImplicitFlowConfiguration;
}());
var AuthConfiguration = /** @class */ (function () {
    /**
     * @param {?} defaultConfig
     */
    function AuthConfiguration(defaultConfig) {
        this.defaultConfig = defaultConfig;
    }
    Object.defineProperty(AuthConfiguration.prototype, "stsServer", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.stsServer;
            }
            return this.defaultConfig.stsServer;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "redirect_url", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.redirect_url;
            }
            return this.defaultConfig.redirect_url;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "silent_redirect_url", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.silent_renew_url;
            }
            return this.defaultConfig.silent_renew_url;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "client_id", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.client_id;
            }
            return this.defaultConfig.client_id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "response_type", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.response_type;
            }
            return this.defaultConfig.response_type;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "scope", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.scope;
            }
            return this.defaultConfig.scope;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "hd_param", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.hd_param;
            }
            return this.defaultConfig.hd_param;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "post_logout_redirect_uri", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.post_logout_redirect_uri;
            }
            return this.defaultConfig.post_logout_redirect_uri;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "start_checksession", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.start_checksession;
            }
            return this.defaultConfig.start_checksession;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "silent_renew", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.silent_renew;
            }
            return this.defaultConfig.silent_renew;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "silent_renew_offset_in_seconds", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.silent_renew_offset_in_seconds;
            }
            return this.defaultConfig.silent_renew_offset_in_seconds;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "post_login_route", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.post_login_route;
            }
            return this.defaultConfig.post_login_route;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "forbidden_route", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.forbidden_route;
            }
            return this.defaultConfig.forbidden_route;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "unauthorized_route", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.unauthorized_route;
            }
            return this.defaultConfig.unauthorized_route;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "auto_userinfo", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.auto_userinfo;
            }
            return this.defaultConfig.auto_userinfo;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "auto_clean_state_after_authentication", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.auto_clean_state_after_authentication;
            }
            return this.defaultConfig.auto_clean_state_after_authentication;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "trigger_authorization_result_event", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.trigger_authorization_result_event;
            }
            return this.defaultConfig.trigger_authorization_result_event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "log_console_warning_active", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.log_console_warning_active;
            }
            return this.defaultConfig.log_console_warning_active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "log_console_debug_active", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.log_console_debug_active;
            }
            return this.defaultConfig.log_console_debug_active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "max_id_token_iat_offset_allowed_in_seconds", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.max_id_token_iat_offset_allowed_in_seconds;
            }
            return this.defaultConfig.max_id_token_iat_offset_allowed_in_seconds;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AuthConfiguration.prototype, "storage", {
        /**
         * @return {?}
         */
        get: function () {
            if (this.openIDImplicitFlowConfiguration) {
                return this.openIDImplicitFlowConfiguration.storage;
            }
            return this.defaultConfig.storage;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} openIDImplicitFlowConfiguration
     * @return {?}
     */
    AuthConfiguration.prototype.init = function (openIDImplicitFlowConfiguration) {
        this.openIDImplicitFlowConfiguration = openIDImplicitFlowConfiguration;
    };
    return AuthConfiguration;
}());
AuthConfiguration.decorators = [
    { type: Injectable },
];
/** @nocollapse */
AuthConfiguration.ctorParameters = function () { return [
    { type: DefaultConfiguration, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Implement this class-interface to create a custom storage.
 * @abstract
 */
var OidcSecurityStorage = /** @class */ (function () {
    function OidcSecurityStorage() {
    }
    return OidcSecurityStorage;
}());
OidcSecurityStorage.decorators = [
    { type: Injectable },
];
var BrowserStorage = /** @class */ (function () {
    /**
     * @param {?} authConfiguration
     */
    function BrowserStorage(authConfiguration) {
        this.authConfiguration = authConfiguration;
        this.hasStorage = typeof Storage !== 'undefined';
    }
    /**
     * @param {?} key
     * @return {?}
     */
    BrowserStorage.prototype.read = function (key) {
        if (this.hasStorage) {
            return JSON.parse(this.authConfiguration.storage.getItem(key));
        }
        return;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    BrowserStorage.prototype.write = function (key, value) {
        if (this.hasStorage) {
            value = value === undefined ? null : value;
            this.authConfiguration.storage.setItem(key, JSON.stringify(value));
        }
    };
    return BrowserStorage;
}());
BrowserStorage.decorators = [
    { type: Injectable },
];
/** @nocollapse */
BrowserStorage.ctorParameters = function () { return [
    { type: AuthConfiguration, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcSecurityCommon = /** @class */ (function () {
    /**
     * @param {?} oidcSecurityStorage
     */
    function OidcSecurityCommon(oidcSecurityStorage) {
        this.oidcSecurityStorage = oidcSecurityStorage;
        this.storage_auth_result = 'authorizationResult';
        this.storage_access_token = 'authorizationData';
        this.storage_id_token = 'authorizationDataIdToken';
        this.storage_is_authorized = '_isAuthorized';
        this.storage_user_data = 'userData';
        this.storage_auth_nonce = 'authNonce';
        this.storage_auth_state_control = 'authStateControl';
        this.storage_session_state = 'session_state';
        this.storage_silent_renew_running = 'storage_silent_renew_running';
        this.storage_custom_request_params = 'storage_custom_request_params';
    }
    Object.defineProperty(OidcSecurityCommon.prototype, "authResult", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_auth_result);
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_auth_result, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "accessToken", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_access_token) || '';
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_access_token, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "idToken", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_id_token) || '';
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_id_token, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "isAuthorized", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_is_authorized);
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_is_authorized, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "userData", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_user_data);
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_user_data, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "authNonce", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_auth_nonce) || '';
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_auth_nonce, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "authStateControl", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_auth_state_control) || '';
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_auth_state_control, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "sessionState", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_session_state);
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_session_state, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "silentRenewRunning", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_silent_renew_running) || '';
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_silent_renew_running, value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OidcSecurityCommon.prototype, "customRequestParams", {
        /**
         * @return {?}
         */
        get: function () {
            return this.retrieve(this.storage_custom_request_params);
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            this.store(this.storage_custom_request_params, value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    OidcSecurityCommon.prototype.setupModule = function () { };
    /**
     * @param {?} key
     * @return {?}
     */
    OidcSecurityCommon.prototype.retrieve = function (key) {
        return this.oidcSecurityStorage.read(key);
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    OidcSecurityCommon.prototype.store = function (key, value) {
        this.oidcSecurityStorage.write(key, value);
    };
    /**
     * @param {?} isRenewProcess
     * @return {?}
     */
    OidcSecurityCommon.prototype.resetStorageData = function (isRenewProcess) {
        if (!isRenewProcess) {
            this.store(this.storage_auth_result, '');
            this.store(this.storage_session_state, '');
            this.store(this.storage_silent_renew_running, '');
            this.store(this.storage_is_authorized, false);
            this.store(this.storage_access_token, '');
            this.store(this.storage_id_token, '');
            this.store(this.storage_user_data, '');
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityCommon.prototype.getAccessToken = function () {
        return this.retrieve(this.storage_access_token);
    };
    /**
     * @return {?}
     */
    OidcSecurityCommon.prototype.getIdToken = function () {
        return this.retrieve(this.storage_id_token);
    };
    return OidcSecurityCommon;
}());
OidcSecurityCommon.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityCommon.ctorParameters = function () { return [
    { type: OidcSecurityStorage, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var ArrayHelperService = /** @class */ (function () {
    function ArrayHelperService() {
    }
    /**
     * @param {?} arr1
     * @param {?} arr2
     * @return {?}
     */
    ArrayHelperService.prototype.arraysEqual = function (arr1, arr2) {
        if (arr1.length !== arr2.length) {
            return false;
        }
        for (var /** @type {?} */ i = arr1.length; i--;) {
            if (arr1[i] !== arr2[i]) {
                return false;
            }
        }
        return true;
    };
    return ArrayHelperService;
}());
ArrayHelperService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
ArrayHelperService.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var TokenHelperService = /** @class */ (function () {
    function TokenHelperService() {
    }
    /**
     * @param {?} dataIdToken
     * @return {?}
     */
    TokenHelperService.prototype.getTokenExpirationDate = function (dataIdToken) {
        if (!dataIdToken.hasOwnProperty('exp')) {
            return new Date();
        }
        var /** @type {?} */ date = new Date(0); // The 0 here is the key, which sets the date to the epoch
        date.setUTCSeconds(dataIdToken.exp);
        return date;
    };
    /**
     * @param {?} token
     * @param {?} encode
     * @return {?}
     */
    TokenHelperService.prototype.getPayloadFromToken = function (token, encode) {
        var /** @type {?} */ data = {};
        if (typeof token !== 'undefined') {
            var /** @type {?} */ encoded = token.split('.')[1];
            if (encode) {
                return encoded;
            }
            data = JSON.parse(this.urlBase64Decode(encoded));
        }
        return data;
    };
    /**
     * @param {?} token
     * @param {?} encode
     * @return {?}
     */
    TokenHelperService.prototype.getHeaderFromToken = function (token, encode) {
        var /** @type {?} */ data = {};
        if (typeof token !== 'undefined') {
            var /** @type {?} */ encoded = token.split('.')[0];
            if (encode) {
                return encoded;
            }
            data = JSON.parse(this.urlBase64Decode(encoded));
        }
        return data;
    };
    /**
     * @param {?} token
     * @param {?} encode
     * @return {?}
     */
    TokenHelperService.prototype.getSignatureFromToken = function (token, encode) {
        var /** @type {?} */ data = {};
        if (typeof token !== 'undefined') {
            var /** @type {?} */ encoded = token.split('.')[2];
            if (encode) {
                return encoded;
            }
            data = JSON.parse(this.urlBase64Decode(encoded));
        }
        return data;
    };
    /**
     * @param {?} str
     * @return {?}
     */
    TokenHelperService.prototype.urlBase64Decode = function (str) {
        var /** @type {?} */ output = str.replace('-', '+').replace('_', '/');
        switch (output.length % 4) {
            case 0:
                break;
            case 2:
                output += '==';
                break;
            case 3:
                output += '=';
                break;
            default:
                throw Error('Illegal base64url string!');
        }
        return window.atob(output);
    };
    return TokenHelperService;
}());
TokenHelperService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
TokenHelperService.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var LoggerService = /** @class */ (function () {
    /**
     * @param {?} authConfiguration
     */
    function LoggerService(authConfiguration) {
        this.authConfiguration = authConfiguration;
    }
    /**
     * @param {?} message
     * @return {?}
     */
    LoggerService.prototype.logError = function (message) {
        console.error(message);
    };
    /**
     * @param {?} message
     * @return {?}
     */
    LoggerService.prototype.logWarning = function (message) {
        if (this.authConfiguration.log_console_warning_active) {
            console.warn(message);
        }
    };
    /**
     * @param {?} message
     * @return {?}
     */
    LoggerService.prototype.logDebug = function (message) {
        if (this.authConfiguration.log_console_debug_active) {
            console.log(message);
        }
    };
    return LoggerService;
}());
LoggerService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LoggerService.ctorParameters = function () { return [
    { type: AuthConfiguration, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcSecurityValidation = /** @class */ (function () {
    /**
     * @param {?} arrayHelperService
     * @param {?} tokenHelperService
     * @param {?} loggerService
     */
    function OidcSecurityValidation(arrayHelperService, tokenHelperService, loggerService) {
        this.arrayHelperService = arrayHelperService;
        this.tokenHelperService = tokenHelperService;
        this.loggerService = loggerService;
    }
    /**
     * @param {?} token
     * @param {?=} offsetSeconds
     * @return {?}
     */
    OidcSecurityValidation.prototype.isTokenExpired = function (token, offsetSeconds) {
        var /** @type {?} */ decoded;
        decoded = this.tokenHelperService.getPayloadFromToken(token, false);
        return !this.validate_id_token_exp_not_expired(decoded, offsetSeconds);
    };
    /**
     * @param {?} decoded_id_token
     * @param {?=} offsetSeconds
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_id_token_exp_not_expired = function (decoded_id_token, offsetSeconds) {
        var /** @type {?} */ tokenExpirationDate = this.tokenHelperService.getTokenExpirationDate(decoded_id_token);
        offsetSeconds = offsetSeconds || 0;
        if (!tokenExpirationDate) {
            return false;
        }
        // Token not expired?
        return (tokenExpirationDate.valueOf() >
            new Date().valueOf() + offsetSeconds * 1000);
    };
    /**
     * @param {?} dataIdToken
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_required_id_token = function (dataIdToken) {
        var /** @type {?} */ validated = true;
        if (!dataIdToken.hasOwnProperty('iss')) {
            validated = false;
            this.loggerService.logWarning('iss is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('sub')) {
            validated = false;
            this.loggerService.logWarning('sub is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('aud')) {
            validated = false;
            this.loggerService.logWarning('aud is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('exp')) {
            validated = false;
            this.loggerService.logWarning('exp is missing, this is required in the id_token');
        }
        if (!dataIdToken.hasOwnProperty('iat')) {
            validated = false;
            this.loggerService.logWarning('iat is missing, this is required in the id_token');
        }
        return validated;
    };
    /**
     * @param {?} dataIdToken
     * @param {?} max_offset_allowed_in_seconds
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_id_token_iat_max_offset = function (dataIdToken, max_offset_allowed_in_seconds) {
        if (!dataIdToken.hasOwnProperty('iat')) {
            return false;
        }
        var /** @type {?} */ dateTime_iat_id_token = new Date(0); // The 0 here is the key, which sets the date to the epoch
        dateTime_iat_id_token.setUTCSeconds(dataIdToken.iat);
        max_offset_allowed_in_seconds = max_offset_allowed_in_seconds || 0;
        if (dateTime_iat_id_token == null) {
            return false;
        }
        this.loggerService.logDebug('validate_id_token_iat_max_offset: ' +
            (new Date().valueOf() - dateTime_iat_id_token.valueOf()) +
            ' < ' +
            max_offset_allowed_in_seconds * 1000);
        return (new Date().valueOf() - dateTime_iat_id_token.valueOf() <
            max_offset_allowed_in_seconds * 1000);
    };
    /**
     * @param {?} dataIdToken
     * @param {?} local_nonce
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_id_token_nonce = function (dataIdToken, local_nonce) {
        if (dataIdToken.nonce !== local_nonce) {
            this.loggerService.logDebug('Validate_id_token_nonce failed, dataIdToken.nonce: ' +
                dataIdToken.nonce +
                ' local_nonce:' +
                local_nonce);
            return false;
        }
        return true;
    };
    /**
     * @param {?} dataIdToken
     * @param {?} authWellKnownEndpoints_issuer
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_id_token_iss = function (dataIdToken, authWellKnownEndpoints_issuer) {
        if (((dataIdToken.iss)) !==
            ((authWellKnownEndpoints_issuer))) {
            this.loggerService.logDebug('Validate_id_token_iss failed, dataIdToken.iss: ' +
                dataIdToken.iss +
                ' authWellKnownEndpoints issuer:' +
                authWellKnownEndpoints_issuer);
            return false;
        }
        return true;
    };
    /**
     * @param {?} dataIdToken
     * @param {?} aud
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_id_token_aud = function (dataIdToken, aud) {
        if (dataIdToken.aud instanceof Array) {
            var /** @type {?} */ result = this.arrayHelperService.arraysEqual(dataIdToken.aud, aud);
            if (!result) {
                this.loggerService.logDebug('Validate_id_token_aud  array failed, dataIdToken.aud: ' +
                    dataIdToken.aud +
                    ' client_id:' +
                    aud);
                return false;
            }
            return true;
        }
        else if (dataIdToken.aud !== aud) {
            this.loggerService.logDebug('Validate_id_token_aud failed, dataIdToken.aud: ' +
                dataIdToken.aud +
                ' client_id:' +
                aud);
            return false;
        }
        return true;
    };
    /**
     * @param {?} state
     * @param {?} local_state
     * @return {?}
     */
    OidcSecurityValidation.prototype.validateStateFromHashCallback = function (state, local_state) {
        if (((state)) !== ((local_state))) {
            this.loggerService.logDebug('ValidateStateFromHashCallback failed, state: ' +
                state +
                ' local_state:' +
                local_state);
            return false;
        }
        return true;
    };
    /**
     * @param {?} id_token_sub
     * @param {?} userdata_sub
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_userdata_sub_id_token = function (id_token_sub, userdata_sub) {
        if (((id_token_sub)) !== ((userdata_sub))) {
            this.loggerService.logDebug('validate_userdata_sub_id_token failed, id_token_sub: ' +
                id_token_sub +
                ' userdata_sub:' +
                userdata_sub);
            return false;
        }
        return true;
    };
    /**
     * @param {?} id_token
     * @param {?} jwtkeys
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_signature_id_token = function (id_token, jwtkeys) {
        if (!jwtkeys || !jwtkeys.keys) {
            return false;
        }
        var /** @type {?} */ header_data = this.tokenHelperService.getHeaderFromToken(id_token, false);
        if (Object.keys(header_data).length === 0 &&
            header_data.constructor === Object) {
            this.loggerService.logWarning('id token has no header data');
            return false;
        }
        var /** @type {?} */ kid = header_data.kid;
        var /** @type {?} */ alg = header_data.alg;
        if ('RS256' !== ((alg))) {
            this.loggerService.logWarning('Only RS256 supported');
            return false;
        }
        var /** @type {?} */ isValid = false;
        if (!header_data.hasOwnProperty('kid')) {
            // exactly 1 key in the jwtkeys and no kid in the Jose header
            // kty	"RSA" use "sig"
            var /** @type {?} */ amountOfMatchingKeys = 0;
            for (var _i = 0, _a = jwtkeys.keys; _i < _a.length; _i++) {
                var key = _a[_i];
                if (((key.kty)) === 'RSA' &&
                    ((key.use)) === 'sig') {
                    amountOfMatchingKeys = amountOfMatchingKeys + 1;
                }
            }
            if (amountOfMatchingKeys === 0) {
                this.loggerService.logWarning('no keys found, incorrect Signature, validation failed for id_token');
                return false;
            }
            else if (amountOfMatchingKeys > 1) {
                this.loggerService.logWarning('no ID Token kid claim in JOSE header and multiple supplied in jwks_uri');
                return false;
            }
            else {
                for (var _b = 0, _c = jwtkeys.keys; _b < _c.length; _b++) {
                    var key = _c[_b];
                    if (((key.kty)) === 'RSA' &&
                        ((key.use)) === 'sig') {
                        var /** @type {?} */ publickey = KEYUTIL.getKey(key);
                        isValid = KJUR.jws.JWS.verify(id_token, publickey, [
                            'RS256'
                        ]);
                        if (!isValid) {
                            this.loggerService.logWarning('incorrect Signature, validation failed for id_token');
                        }
                        return isValid;
                    }
                }
            }
        }
        else {
            // kid in the Jose header of id_token
            for (var _d = 0, _e = jwtkeys.keys; _d < _e.length; _d++) {
                var key = _e[_d];
                if (((key.kid)) === ((kid))) {
                    var /** @type {?} */ publickey = KEYUTIL.getKey(key);
                    isValid = KJUR.jws.JWS.verify(id_token, publickey, [
                        'RS256'
                    ]);
                    if (!isValid) {
                        this.loggerService.logWarning('incorrect Signature, validation failed for id_token');
                    }
                    return isValid;
                }
            }
        }
        return isValid;
    };
    /**
     * @param {?} response_type
     * @return {?}
     */
    OidcSecurityValidation.prototype.config_validate_response_type = function (response_type) {
        if (response_type === 'id_token token' ||
            response_type === 'id_token') {
            return true;
        }
        this.loggerService.logWarning('module configure incorrect, invalid response_type:' + response_type);
        return false;
    };
    /**
     * @param {?} access_token
     * @param {?} at_hash
     * @return {?}
     */
    OidcSecurityValidation.prototype.validate_id_token_at_hash = function (access_token, at_hash) {
        this.loggerService.logDebug('From the server:' + at_hash);
        var /** @type {?} */ testdata = this.generate_at_hash('' + access_token);
        this.loggerService.logDebug('client validation not decoded:' + testdata);
        if (testdata === ((at_hash))) {
            return true; // isValid;
        }
        else {
            var /** @type {?} */ testValue = this.generate_at_hash('' + decodeURIComponent(access_token));
            this.loggerService.logDebug('-gen access--' + testValue);
            if (testValue === ((at_hash))) {
                return true; // isValid
            }
        }
        return false;
    };
    /**
     * @param {?} access_token
     * @return {?}
     */
    OidcSecurityValidation.prototype.generate_at_hash = function (access_token) {
        var /** @type {?} */ hash = KJUR.crypto.Util.hashString(access_token, 'sha256');
        var /** @type {?} */ first128bits = hash.substr(0, hash.length / 2);
        var /** @type {?} */ testdata = hextob64u(first128bits);
        return testdata;
    };
    return OidcSecurityValidation;
}());
OidcSecurityValidation.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityValidation.ctorParameters = function () { return [
    { type: ArrayHelperService, },
    { type: TokenHelperService, },
    { type: LoggerService, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var StateValidationService = /** @class */ (function () {
    /**
     * @param {?} authConfiguration
     * @param {?} oidcSecurityCommon
     * @param {?} oidcSecurityValidation
     * @param {?} tokenHelperService
     * @param {?} loggerService
     */
    function StateValidationService(authConfiguration, oidcSecurityCommon, oidcSecurityValidation, tokenHelperService, loggerService) {
        this.authConfiguration = authConfiguration;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.oidcSecurityValidation = oidcSecurityValidation;
        this.tokenHelperService = tokenHelperService;
        this.loggerService = loggerService;
    }
    /**
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    StateValidationService.prototype.setupModule = function (authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
    };
    /**
     * @param {?} result
     * @param {?} jwtKeys
     * @return {?}
     */
    StateValidationService.prototype.validateState = function (result, jwtKeys) {
        var /** @type {?} */ toReturn = new ValidateStateResult('', '', false, {});
        if (!this.oidcSecurityValidation.validateStateFromHashCallback(result.state, this.oidcSecurityCommon.authStateControl)) {
            this.loggerService.logWarning('authorizedCallback incorrect state');
            return toReturn;
        }
        if (this.authConfiguration.response_type === 'id_token token') {
            toReturn.access_token = result.access_token;
        }
        toReturn.id_token = result.id_token;
        toReturn.decoded_id_token = this.tokenHelperService.getPayloadFromToken(toReturn.id_token, false);
        if (!this.oidcSecurityValidation.validate_signature_id_token(toReturn.id_token, jwtKeys)) {
            this.loggerService.logDebug('authorizedCallback Signature validation failed id_token');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_nonce(toReturn.decoded_id_token, this.oidcSecurityCommon.authNonce)) {
            this.loggerService.logWarning('authorizedCallback incorrect nonce');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_required_id_token(toReturn.decoded_id_token)) {
            this.loggerService.logDebug('authorizedCallback Validation, one of the REQUIRED properties missing from id_token');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_iat_max_offset(toReturn.decoded_id_token, this.authConfiguration
            .max_id_token_iat_offset_allowed_in_seconds)) {
            this.loggerService.logWarning('authorizedCallback Validation, iat rejected id_token was issued too far away from the current time');
            return toReturn;
        }
        if (this.authWellKnownEndpoints) {
            if (!this.oidcSecurityValidation.validate_id_token_iss(toReturn.decoded_id_token, this.authWellKnownEndpoints.issuer)) {
                this.loggerService.logWarning('authorizedCallback incorrect iss does not match authWellKnownEndpoints issuer');
                return toReturn;
            }
        }
        else {
            this.loggerService.logWarning('authWellKnownEndpoints is undefined');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_aud(toReturn.decoded_id_token, this.authConfiguration.client_id)) {
            this.loggerService.logWarning('authorizedCallback incorrect aud');
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_exp_not_expired(toReturn.decoded_id_token)) {
            this.loggerService.logWarning('authorizedCallback token expired');
            return toReturn;
        }
        // flow id_token token
        if (this.authConfiguration.response_type !== 'id_token token') {
            toReturn.authResponseIsValid = true;
            this.handleSuccessfulValidation();
            return toReturn;
        }
        if (!this.oidcSecurityValidation.validate_id_token_at_hash(toReturn.access_token, toReturn.decoded_id_token.at_hash) ||
            !toReturn.access_token) {
            this.loggerService.logWarning('authorizedCallback incorrect at_hash');
            return toReturn;
        }
        toReturn.authResponseIsValid = true;
        this.handleSuccessfulValidation();
        return toReturn;
    };
    /**
     * @return {?}
     */
    StateValidationService.prototype.handleSuccessfulValidation = function () {
        this.oidcSecurityCommon.authNonce = '';
        if (this.authConfiguration.auto_clean_state_after_authentication) {
            this.oidcSecurityCommon.authStateControl = '';
        }
        this.loggerService.logDebug('AuthorizedCallback token(s) validated, continue');
    };
    return StateValidationService;
}());
StateValidationService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
StateValidationService.ctorParameters = function () { return [
    { type: AuthConfiguration, },
    { type: OidcSecurityCommon, },
    { type: OidcSecurityValidation, },
    { type: TokenHelperService, },
    { type: LoggerService, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcSecurityCheckSession = /** @class */ (function () {
    /**
     * @param {?} authConfiguration
     * @param {?} oidcSecurityCommon
     * @param {?} loggerService
     * @param {?} zone
     */
    function OidcSecurityCheckSession(authConfiguration, oidcSecurityCommon, loggerService, zone) {
        this.authConfiguration = authConfiguration;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.loggerService = loggerService;
        this.zone = zone;
        this.onCheckSessionChanged = new EventEmitter(true);
    }
    /**
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.setupModule = function (authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
    };
    /**
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.doesSessionExist = function () {
        var /** @type {?} */ existsparent = undefined;
        try {
            var /** @type {?} */ parentdoc = window.parent.document;
            if (!parentdoc) {
                throw new Error('Unaccessible');
            }
            existsparent = parentdoc.getElementById('myiFrameForCheckSession');
        }
        catch (e) {
            // not accessible
        }
        var /** @type {?} */ exists = window.document.getElementById('myiFrameForCheckSession');
        if (existsparent) {
            this.sessionIframe = existsparent;
        }
        else if (exists) {
            this.sessionIframe = exists;
        }
        if (existsparent || exists) {
            return true;
        }
        return false;
    };
    /**
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.init = function () {
        var _this = this;
        this.sessionIframe = window.document.createElement('iframe');
        this.sessionIframe.id = 'myiFrameForCheckSession';
        this.loggerService.logDebug(this.sessionIframe);
        this.sessionIframe.style.display = 'none';
        window.document.body.appendChild(this.sessionIframe);
        if (this.authWellKnownEndpoints) {
            this.sessionIframe.src = this.authWellKnownEndpoints.check_session_iframe;
        }
        else {
            this.loggerService.logWarning('init check session: authWellKnownEndpoints is undefined');
        }
        this.iframeMessageEvent = this.messageHandler.bind(this);
        window.addEventListener('message', this.iframeMessageEvent, false);
        return Observable.create(function (observer) {
            _this.sessionIframe.onload = function () {
                observer.next(_this);
                observer.complete();
            };
        });
    };
    /**
     * @param {?} clientId
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.startCheckingSession = function (clientId) {
        if (!this._scheduledHeartBeat) {
            this.pollServerSession(clientId);
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.stopCheckingSession = function () {
        if (this._scheduledHeartBeat) {
            clearTimeout(this._scheduledHeartBeat);
            this._scheduledHeartBeat = null;
        }
    };
    /**
     * @param {?} clientId
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.pollServerSession = function (clientId) {
        var _this = this;
        var /** @type {?} */ _pollServerSessionRecur = function () {
            if (_this.sessionIframe && clientId) {
                _this.loggerService.logDebug(_this.sessionIframe);
                var /** @type {?} */ session_state = _this.oidcSecurityCommon.sessionState;
                if (session_state) {
                    _this.sessionIframe.contentWindow.postMessage(clientId + ' ' + session_state, _this.authConfiguration.stsServer);
                }
            }
            else {
                _this.loggerService.logWarning('OidcSecurityCheckSession pollServerSession sessionIframe does not exist');
                _this.loggerService.logDebug(clientId);
                _this.loggerService.logDebug(_this.sessionIframe);
                // this.init();
            }
            _this._scheduledHeartBeat = setTimeout(_pollServerSessionRecur, 3000);
        };
        this.zone.runOutsideAngular(function () {
            _this._scheduledHeartBeat = setTimeout(_pollServerSessionRecur, 3000);
        });
    };
    /**
     * @param {?} e
     * @return {?}
     */
    OidcSecurityCheckSession.prototype.messageHandler = function (e) {
        if (this.sessionIframe &&
            e.origin === this.authConfiguration.stsServer &&
            e.source === this.sessionIframe.contentWindow) {
            if (e.data === 'error') {
                this.loggerService.logWarning('error from checksession messageHandler');
            }
            else if (e.data === 'changed') {
                this.onCheckSessionChanged.emit();
            }
            else {
                this.loggerService.logDebug(e.data + ' from checksession messageHandler');
            }
        }
    };
    return OidcSecurityCheckSession;
}());
OidcSecurityCheckSession.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityCheckSession.ctorParameters = function () { return [
    { type: AuthConfiguration, },
    { type: OidcSecurityCommon, },
    { type: LoggerService, },
    { type: NgZone, },
]; };
OidcSecurityCheckSession.propDecorators = {
    "onCheckSessionChanged": [{ type: Output },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcSecuritySilentRenew = /** @class */ (function () {
    /**
     * @param {?} loggerService
     */
    function OidcSecuritySilentRenew(loggerService) {
        this.loggerService = loggerService;
    }
    /**
     * @return {?}
     */
    OidcSecuritySilentRenew.prototype.initRenew = function () {
        var /** @type {?} */ existsparent = undefined;
        try {
            var /** @type {?} */ parentdoc = window.parent.document;
            if (!parentdoc) {
                throw new Error('Unaccessible');
            }
            existsparent = parentdoc.getElementById('myiFrameForSilentRenew');
        }
        catch (e) {
            // not accessible
        }
        var /** @type {?} */ exists = window.document.getElementById('myiFrameForSilentRenew');
        if (existsparent) {
            this.sessionIframe = existsparent;
        }
        else if (exists) {
            this.sessionIframe = exists;
        }
        if (!exists && !existsparent) {
            this.sessionIframe = window.document.createElement('iframe');
            this.sessionIframe.id = 'myiFrameForSilentRenew';
            this.loggerService.logDebug(this.sessionIframe);
            this.sessionIframe.style.display = 'none';
            window.document.body.appendChild(this.sessionIframe);
        }
    };
    /**
     * @param {?} url
     * @return {?}
     */
    OidcSecuritySilentRenew.prototype.startRenew = function (url) {
        var _this = this;
        var /** @type {?} */ existsparent = undefined;
        try {
            var /** @type {?} */ parentdoc = window.parent.document;
            if (!parentdoc) {
                throw new Error('Unaccessible');
            }
            existsparent = parentdoc.getElementById('myiFrameForSilentRenew');
        }
        catch (e) {
            // not accessible
        }
        var /** @type {?} */ exists = window.document.getElementById('myiFrameForSilentRenew');
        if (existsparent) {
            this.sessionIframe = existsparent;
        }
        else if (exists) {
            this.sessionIframe = exists;
        }
        this.loggerService.logDebug('startRenew for URL:' + url);
        this.sessionIframe.src = url;
        return Observable.create(function (observer) {
            _this.sessionIframe.onload = function () {
                observer.next(_this);
                observer.complete();
            };
        });
    };
    return OidcSecuritySilentRenew;
}());
OidcSecuritySilentRenew.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecuritySilentRenew.ctorParameters = function () { return [
    { type: LoggerService, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcDataService = /** @class */ (function () {
    /**
     * @param {?} httpClient
     */
    function OidcDataService(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * @template T
     * @param {?} url
     * @return {?}
     */
    OidcDataService.prototype.getWellknownEndpoints = function (url) {
        var /** @type {?} */ headers = new HttpHeaders();
        headers = headers.set('Accept', 'application/json');
        return this.httpClient.get(url, {
            headers: headers
        });
    };
    /**
     * @template T
     * @param {?} url
     * @param {?} token
     * @return {?}
     */
    OidcDataService.prototype.getIdentityUserData = function (url, token) {
        var /** @type {?} */ headers = new HttpHeaders();
        headers = headers.set('Accept', 'application/json');
        headers = headers.set('Authorization', 'Bearer ' + decodeURIComponent(token));
        return this.httpClient.get(url, {
            headers: headers
        });
    };
    /**
     * @template T
     * @param {?} url
     * @return {?}
     */
    OidcDataService.prototype.get = function (url) {
        var /** @type {?} */ headers = new HttpHeaders();
        headers = headers.set('Accept', 'application/json');
        return this.httpClient.get(url, {
            headers: headers
        });
    };
    return OidcDataService;
}());
OidcDataService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcDataService.ctorParameters = function () { return [
    { type: HttpClient, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcSecurityUserService = /** @class */ (function () {
    /**
     * @param {?} oidcDataService
     * @param {?} oidcSecurityCommon
     * @param {?} loggerService
     */
    function OidcSecurityUserService(oidcDataService, oidcSecurityCommon, loggerService) {
        this.oidcDataService = oidcDataService;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.loggerService = loggerService;
        this.userData = '';
    }
    /**
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    OidcSecurityUserService.prototype.setupModule = function (authWellKnownEndpoints) {
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
    };
    /**
     * @return {?}
     */
    OidcSecurityUserService.prototype.initUserData = function () {
        var _this = this;
        return this.getIdentityUserData().pipe(map(function (data) { return (_this.userData = data); }));
    };
    /**
     * @return {?}
     */
    OidcSecurityUserService.prototype.getUserData = function () {
        if (!this.userData) {
            throw Error('UserData is not set!');
        }
        return this.userData;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    OidcSecurityUserService.prototype.setUserData = function (value) {
        this.userData = value;
    };
    /**
     * @return {?}
     */
    OidcSecurityUserService.prototype.getIdentityUserData = function () {
        var /** @type {?} */ token = this.oidcSecurityCommon.getAccessToken();
        if (this.authWellKnownEndpoints) {
            return this.oidcDataService.getIdentityUserData(this.authWellKnownEndpoints.userinfo_endpoint, token);
        }
        else {
            this.loggerService.logWarning('init check session: authWellKnownEndpoints is undefined');
        }
        return this.oidcDataService.getIdentityUserData('undefined', token);
    };
    return OidcSecurityUserService;
}());
OidcSecurityUserService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityUserService.ctorParameters = function () { return [
    { type: OidcDataService, },
    { type: OidcSecurityCommon, },
    { type: LoggerService, },
]; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var UriEncoder = /** @class */ (function () {
    function UriEncoder() {
    }
    /**
     * @param {?} key
     * @return {?}
     */
    UriEncoder.prototype.encodeKey = function (key) {
        return encodeURIComponent(key);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    UriEncoder.prototype.encodeValue = function (value) {
        return encodeURIComponent(value);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    UriEncoder.prototype.decodeKey = function (key) {
        return decodeURIComponent(key);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    UriEncoder.prototype.decodeValue = function (value) {
        return decodeURIComponent(value);
    };
    return UriEncoder;
}());
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OidcSecurityService = /** @class */ (function () {
    /**
     * @param {?} platformId
     * @param {?} oidcDataService
     * @param {?} stateValidationService
     * @param {?} authConfiguration
     * @param {?} router
     * @param {?} oidcSecurityCheckSession
     * @param {?} oidcSecuritySilentRenew
     * @param {?} oidcSecurityUserService
     * @param {?} oidcSecurityCommon
     * @param {?} oidcSecurityValidation
     * @param {?} tokenHelperService
     * @param {?} loggerService
     * @param {?} zone
     */
    function OidcSecurityService(platformId, oidcDataService, stateValidationService, authConfiguration, router, oidcSecurityCheckSession, oidcSecuritySilentRenew, oidcSecurityUserService, oidcSecurityCommon, oidcSecurityValidation, tokenHelperService, loggerService, zone) {
        this.platformId = platformId;
        this.oidcDataService = oidcDataService;
        this.stateValidationService = stateValidationService;
        this.authConfiguration = authConfiguration;
        this.router = router;
        this.oidcSecurityCheckSession = oidcSecurityCheckSession;
        this.oidcSecuritySilentRenew = oidcSecuritySilentRenew;
        this.oidcSecurityUserService = oidcSecurityUserService;
        this.oidcSecurityCommon = oidcSecurityCommon;
        this.oidcSecurityValidation = oidcSecurityValidation;
        this.tokenHelperService = tokenHelperService;
        this.loggerService = loggerService;
        this.zone = zone;
        this.onModuleSetup = new EventEmitter();
        this.onAuthorizationResult = new EventEmitter();
        this.onCheckSessionChanged = new EventEmitter();
        this.checkSessionChanged = false;
        this.moduleSetup = false;
        this._isAuthorized = new BehaviorSubject(false);
        this._isAuthorizedValue = false;
        this._userData = new BehaviorSubject('');
        this.authWellKnownEndpointsLoaded = false;
        this.runTokenValidationRunning = false;
    }
    /**
     * @param {?} openIDImplicitFlowConfiguration
     * @param {?} authWellKnownEndpoints
     * @return {?}
     */
    OidcSecurityService.prototype.setupModule = function (openIDImplicitFlowConfiguration, authWellKnownEndpoints) {
        var _this = this;
        this.authWellKnownEndpoints = Object.assign({}, authWellKnownEndpoints);
        this.authConfiguration.init(openIDImplicitFlowConfiguration);
        this.stateValidationService.setupModule(authWellKnownEndpoints);
        this.oidcSecurityCheckSession.setupModule(authWellKnownEndpoints);
        this.oidcSecurityUserService.setupModule(authWellKnownEndpoints);
        this.oidcSecurityCheckSession.onCheckSessionChanged.subscribe(function () {
            _this.loggerService.logDebug('onCheckSessionChanged');
            _this.checkSessionChanged = true;
            _this.onCheckSessionChanged.emit(_this.checkSessionChanged);
        });
        this._userData.subscribe(function () {
            _this.onUserDataChanged();
        });
        var /** @type {?} */ userData = this.oidcSecurityCommon.userData;
        if (userData) {
            this.setUserData(userData);
        }
        var /** @type {?} */ isAuthorized = this.oidcSecurityCommon.isAuthorized;
        if (isAuthorized) {
            this.loggerService.logDebug('IsAuthorized setup module');
            this.loggerService.logDebug(this.oidcSecurityCommon.idToken);
            if (this.oidcSecurityValidation.isTokenExpired(this.oidcSecurityCommon.idToken, this.authConfiguration.silent_renew_offset_in_seconds)) {
                this.loggerService.logDebug('IsAuthorized setup module; id_token isTokenExpired');
            }
            else {
                this.loggerService.logDebug('IsAuthorized setup module; id_token is valid');
                this.setIsAuthorized(isAuthorized);
                this.runTokenValidation();
            }
        }
        this.loggerService.logDebug('STS server: ' + this.authConfiguration.stsServer);
        if (isPlatformBrowser(this.platformId)) {
            // Client only code.
            this.moduleSetup = true;
            this.onModuleSetup.emit();
            if (this.authConfiguration.silent_renew) {
                this.oidcSecuritySilentRenew.initRenew();
                // Support authorization via DOM events.
                // Deregister if OidcSecurityService.setupModule is called again by any instance.
                //      We only ever want the latest setup service to be reacting to this event.
                this.boundSilentRenewEvent = this.silentRenewEventHandler.bind(this);
                var /** @type {?} */ instanceId_1 = Math.random();
                var /** @type {?} */ boundSilentRenewInitEvent_1 = (function (e) {
                    if (e.detail !== instanceId_1) {
                        window.removeEventListener('oidc-silent-renew-message', _this.boundSilentRenewEvent);
                        window.removeEventListener('oidc-silent-renew-init', boundSilentRenewInitEvent_1);
                    }
                }).bind(this);
                window.addEventListener('oidc-silent-renew-init', boundSilentRenewInitEvent_1, false);
                window.addEventListener('oidc-silent-renew-message', this.boundSilentRenewEvent, false);
                window.dispatchEvent(new CustomEvent('oidc-silent-renew-init', { detail: instanceId_1 }));
            }
            if (this.authConfiguration.start_checksession &&
                !this.oidcSecurityCheckSession.doesSessionExist()) {
                this.oidcSecurityCheckSession.init().subscribe(function () {
                    _this.oidcSecurityCheckSession.pollServerSession(_this.authConfiguration.client_id);
                });
            }
        }
        else {
            this.moduleSetup = true;
            this.onModuleSetup.emit();
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.getUserData = function () {
        return this._userData.asObservable();
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.getIsAuthorized = function () {
        return this._isAuthorized.asObservable();
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.getToken = function () {
        if (!this._isAuthorizedValue) {
            return '';
        }
        var /** @type {?} */ token = this.oidcSecurityCommon.getAccessToken();
        return decodeURIComponent(token);
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.getIdToken = function () {
        if (!this._isAuthorizedValue) {
            return '';
        }
        var /** @type {?} */ token = this.oidcSecurityCommon.getIdToken();
        return decodeURIComponent(token);
    };
    /**
     * @param {?=} encode
     * @return {?}
     */
    OidcSecurityService.prototype.getPayloadFromIdToken = function (encode) {
        if (encode === void 0) { encode = false; }
        var /** @type {?} */ token = this.getIdToken();
        return this.tokenHelperService.getPayloadFromToken(token, encode);
    };
    /**
     * @param {?} state
     * @return {?}
     */
    OidcSecurityService.prototype.setState = function (state) {
        this.oidcSecurityCommon.authStateControl = state;
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.getState = function () {
        return this.oidcSecurityCommon.authStateControl;
    };
    /**
     * @param {?} params
     * @return {?}
     */
    OidcSecurityService.prototype.setCustomRequestParameters = function (params) {
        this.oidcSecurityCommon.customRequestParams = params;
    };
    /**
     * @param {?=} urlHandler
     * @return {?}
     */
    OidcSecurityService.prototype.authorize = function (urlHandler) {
        if (this.authWellKnownEndpoints) {
            this.authWellKnownEndpointsLoaded = true;
        }
        if (!this.authWellKnownEndpointsLoaded) {
            this.loggerService.logError('Well known endpoints must be loaded before user can login!');
            return;
        }
        if (!this.oidcSecurityValidation.config_validate_response_type(this.authConfiguration.response_type)) {
            // invalid response_type
            return;
        }
        this.resetAuthorizationData(false);
        this.loggerService.logDebug('BEGIN Authorize, no auth data');
        var /** @type {?} */ state = this.oidcSecurityCommon.authStateControl;
        if (!state) {
            state = Date.now() + '' + Math.random();
            this.oidcSecurityCommon.authStateControl = state;
        }
        var /** @type {?} */ nonce = 'N' + Math.random() + '' + Date.now();
        this.oidcSecurityCommon.authNonce = nonce;
        this.loggerService.logDebug('AuthorizedController created. local state: ' +
            this.oidcSecurityCommon.authStateControl);
        if (this.authWellKnownEndpoints) {
            var /** @type {?} */ url = this.createAuthorizeUrl(this.authConfiguration.redirect_url, nonce, state, this.authWellKnownEndpoints.authorization_endpoint);
            if (urlHandler) {
                urlHandler(url);
            }
            else {
                window.location.href = url;
            }
        }
        else {
            this.loggerService.logError('authWellKnownEndpoints is undefined');
        }
    };
    /**
     * @param {?=} hash
     * @return {?}
     */
    OidcSecurityService.prototype.authorizedCallback = function (hash) {
        var _this = this;
        var /** @type {?} */ silentRenew = this.oidcSecurityCommon.silentRenewRunning;
        var /** @type {?} */ isRenewProcess = silentRenew === 'running';
        this.loggerService.logDebug('BEGIN authorizedCallback, no auth data');
        this.resetAuthorizationData(isRenewProcess);
        hash = hash || window.location.hash.substr(1);
        var /** @type {?} */ result = hash
            .split('&')
            .reduce(function (resultData, item) {
            var /** @type {?} */ parts = item.split('=');
            resultData[parts[0]] = parts[1];
            return resultData;
        }, {});
        this.oidcSecurityCommon.authResult = result;
        this.loggerService.logDebug(result);
        this.loggerService.logDebug('authorizedCallback created, begin token validation');
        this.getSigningKeys().subscribe(function (jwtKeys) {
            var /** @type {?} */ validationResult = _this.getValidatedStateResult(result, jwtKeys);
            if (validationResult.authResponseIsValid) {
                _this.setAuthorizationData(validationResult.access_token, validationResult.id_token);
                _this.oidcSecurityCommon.silentRenewRunning = '';
                if (_this.authConfiguration.auto_userinfo) {
                    _this.getUserinfo(isRenewProcess, result, validationResult.id_token, validationResult.decoded_id_token).subscribe(function (response) {
                        if (response) {
                            _this.onAuthorizationResult.emit(AuthorizationResult.authorized);
                            if (!_this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                                _this.router.navigate([
                                    _this.authConfiguration.post_login_route
                                ]);
                            }
                        }
                        else {
                            _this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
                            if (!_this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                                _this.router.navigate([
                                    _this.authConfiguration.unauthorized_route
                                ]);
                            }
                        }
                    });
                }
                else {
                    if (!isRenewProcess) {
                        // userData is set to the id_token decoded, auto get user data set to false
                        _this.oidcSecurityUserService.setUserData(validationResult.decoded_id_token);
                        _this.setUserData(_this.oidcSecurityUserService.getUserData());
                        _this.runTokenValidation();
                    }
                    _this.onAuthorizationResult.emit(AuthorizationResult.authorized);
                    if (!_this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                        _this.router.navigate([
                            _this.authConfiguration.post_login_route
                        ]);
                    }
                }
            }
            else {
                // something went wrong
                _this.loggerService.logWarning('authorizedCallback, token(s) validation failed, resetting');
                _this.loggerService.logWarning(window.location.hash);
                _this.resetAuthorizationData(false);
                _this.oidcSecurityCommon.silentRenewRunning = '';
                _this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
                if (!_this.authConfiguration.trigger_authorization_result_event && !isRenewProcess) {
                    _this.router.navigate([
                        _this.authConfiguration.unauthorized_route
                    ]);
                }
            }
        }, function (err) {
            /* Something went wrong while getting signing key */
            _this.loggerService.logWarning('Failed to retreive siging key with error: ' + JSON.stringify(err));
            _this.oidcSecurityCommon.silentRenewRunning = '';
        });
    };
    /**
     * @param {?=} isRenewProcess
     * @param {?=} result
     * @param {?=} id_token
     * @param {?=} decoded_id_token
     * @return {?}
     */
    OidcSecurityService.prototype.getUserinfo = function (isRenewProcess, result, id_token, decoded_id_token) {
        var _this = this;
        if (isRenewProcess === void 0) { isRenewProcess = false; }
        result = result ? result : this.oidcSecurityCommon.authResult;
        id_token = id_token ? id_token : this.oidcSecurityCommon.idToken;
        decoded_id_token = decoded_id_token
            ? decoded_id_token
            : this.tokenHelperService.getPayloadFromToken(id_token, false);
        return new Observable(function (observer) {
            // flow id_token token
            if (_this.authConfiguration.response_type === 'id_token token') {
                if (isRenewProcess) {
                    _this.oidcSecurityCommon.sessionState = result.session_state;
                    observer.next(true);
                    observer.complete();
                }
                else {
                    _this.oidcSecurityUserService
                        .initUserData()
                        .subscribe(function () {
                        _this.loggerService.logDebug('authorizedCallback id_token token flow');
                        var /** @type {?} */ userData = _this.oidcSecurityUserService.getUserData();
                        if (_this.oidcSecurityValidation.validate_userdata_sub_id_token(decoded_id_token.sub, userData.sub)) {
                            _this.setUserData(userData);
                            _this.loggerService.logDebug(_this.oidcSecurityCommon.accessToken);
                            _this.loggerService.logDebug(_this.oidcSecurityUserService.getUserData());
                            _this.oidcSecurityCommon.sessionState =
                                result.session_state;
                            _this.runTokenValidation();
                            observer.next(true);
                        }
                        else {
                            // something went wrong, userdata sub does not match that from id_token
                            _this.loggerService.logWarning('authorizedCallback, User data sub does not match sub in id_token');
                            _this.loggerService.logDebug('authorizedCallback, token(s) validation failed, resetting');
                            _this.resetAuthorizationData(false);
                            observer.next(false);
                        }
                        observer.complete();
                    });
                }
            }
            else {
                // flow id_token
                _this.loggerService.logDebug('authorizedCallback id_token flow');
                _this.loggerService.logDebug(_this.oidcSecurityCommon.accessToken);
                // userData is set to the id_token decoded. No access_token.
                _this.oidcSecurityUserService.setUserData(decoded_id_token);
                _this.setUserData(_this.oidcSecurityUserService.getUserData());
                _this.oidcSecurityCommon.sessionState = result.session_state;
                if (!isRenewProcess) {
                    _this.runTokenValidation();
                }
                observer.next(true);
                observer.complete();
            }
        });
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.logoff = function () {
        // /connect/endsession?id_token_hint=...&post_logout_redirect_uri=https://myapp.com
        this.loggerService.logDebug('BEGIN Authorize, no auth data');
        if (this.authWellKnownEndpoints) {
            if (this.authWellKnownEndpoints.end_session_endpoint) {
                var /** @type {?} */ end_session_endpoint = this.authWellKnownEndpoints
                    .end_session_endpoint;
                var /** @type {?} */ id_token_hint = this.oidcSecurityCommon.idToken;
                var /** @type {?} */ url = this.createEndSessionUrl(end_session_endpoint, id_token_hint);
                this.resetAuthorizationData(false);
                if (this.authConfiguration.start_checksession &&
                    this.checkSessionChanged) {
                    this.loggerService.logDebug('only local login cleaned up, server session has changed');
                }
                else {
                    window.location.href = url;
                }
            }
            else {
                this.resetAuthorizationData(false);
                this.loggerService.logDebug('only local login cleaned up, no end_session_endpoint');
            }
        }
        else {
            this.loggerService.logWarning('authWellKnownEndpoints is undefined');
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.refreshSession = function () {
        this.loggerService.logDebug('BEGIN refresh session Authorize');
        var /** @type {?} */ state = this.oidcSecurityCommon.authStateControl;
        if (state === '' || state === null) {
            state = Date.now() + '' + Math.random();
            this.oidcSecurityCommon.authStateControl = state;
        }
        var /** @type {?} */ nonce = 'N' + Math.random() + '' + Date.now();
        this.oidcSecurityCommon.authNonce = nonce;
        this.loggerService.logDebug('RefreshSession created. adding myautostate: ' +
            this.oidcSecurityCommon.authStateControl);
        var /** @type {?} */ url = '';
        if (this.authWellKnownEndpoints) {
            url = this.createAuthorizeUrl(this.authConfiguration.silent_redirect_url, nonce, state, this.authWellKnownEndpoints.authorization_endpoint, 'none');
        }
        else {
            this.loggerService.logWarning('authWellKnownEndpoints is undefined');
        }
        this.oidcSecurityCommon.silentRenewRunning = 'running';
        return this.oidcSecuritySilentRenew.startRenew(url);
    };
    /**
     * @param {?} error
     * @return {?}
     */
    OidcSecurityService.prototype.handleError = function (error) {
        this.loggerService.logError(error);
        if (error.status === 403 || error.status === '403') {
            if (this.authConfiguration.trigger_authorization_result_event) {
                this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
            }
            else {
                this.router.navigate([this.authConfiguration.forbidden_route]);
            }
        }
        else if (error.status === 401 || error.status === '401') {
            var /** @type {?} */ silentRenew = this.oidcSecurityCommon.silentRenewRunning;
            this.resetAuthorizationData(!!silentRenew);
            if (this.authConfiguration.trigger_authorization_result_event) {
                this.onAuthorizationResult.emit(AuthorizationResult.unauthorized);
            }
            else {
                this.router.navigate([
                    this.authConfiguration.unauthorized_route
                ]);
            }
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.startCheckingSilentRenew = function () {
        this.runTokenValidation();
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.stopCheckingSilentRenew = function () {
        if (this._scheduledHeartBeat) {
            clearTimeout(this._scheduledHeartBeat);
            this._scheduledHeartBeat = null;
            this.runTokenValidationRunning = false;
        }
    };
    /**
     * @param {?} result
     * @param {?} jwtKeys
     * @return {?}
     */
    OidcSecurityService.prototype.getValidatedStateResult = function (result, jwtKeys) {
        if (result.error) {
            return new ValidateStateResult('', '', false, {});
        }
        return this.stateValidationService.validateState(result, jwtKeys);
    };
    /**
     * @param {?} userData
     * @return {?}
     */
    OidcSecurityService.prototype.setUserData = function (userData) {
        this.oidcSecurityCommon.userData = userData;
        this._userData.next(userData);
    };
    /**
     * @param {?} isAuthorized
     * @return {?}
     */
    OidcSecurityService.prototype.setIsAuthorized = function (isAuthorized) {
        this._isAuthorizedValue = isAuthorized;
        this._isAuthorized.next(isAuthorized);
    };
    /**
     * @param {?} access_token
     * @param {?} id_token
     * @return {?}
     */
    OidcSecurityService.prototype.setAuthorizationData = function (access_token, id_token) {
        if (this.oidcSecurityCommon.accessToken !== '') {
            this.oidcSecurityCommon.accessToken = '';
        }
        this.loggerService.logDebug(access_token);
        this.loggerService.logDebug(id_token);
        this.loggerService.logDebug('storing to storage, getting the roles');
        this.oidcSecurityCommon.accessToken = access_token;
        this.oidcSecurityCommon.idToken = id_token;
        this.setIsAuthorized(true);
        this.oidcSecurityCommon.isAuthorized = true;
    };
    /**
     * @param {?} redirect_url
     * @param {?} nonce
     * @param {?} state
     * @param {?} authorization_endpoint
     * @param {?=} prompt
     * @return {?}
     */
    OidcSecurityService.prototype.createAuthorizeUrl = function (redirect_url, nonce, state, authorization_endpoint, prompt) {
        var /** @type {?} */ urlParts = authorization_endpoint.split('?');
        var /** @type {?} */ authorizationUrl = urlParts[0];
        var /** @type {?} */ params = new HttpParams({
            fromString: urlParts[1],
            encoder: new UriEncoder()
        });
        params = params.set('client_id', this.authConfiguration.client_id);
        params = params.append('redirect_uri', redirect_url);
        params = params.append('response_type', this.authConfiguration.response_type);
        params = params.append('scope', this.authConfiguration.scope);
        params = params.append('nonce', nonce);
        params = params.append('state', state);
        if (prompt) {
            params = params.append('prompt', prompt);
        }
        if (this.authConfiguration.hd_param) {
            params = params.append('hd', this.authConfiguration.hd_param);
        }
        var /** @type {?} */ customParams = Object.assign({}, this.oidcSecurityCommon.customRequestParams);
        Object.keys(customParams).forEach(function (key) {
            params = params.append(key, customParams[key].toString());
        });
        return authorizationUrl + "?" + params;
    };
    /**
     * @param {?} end_session_endpoint
     * @param {?} id_token_hint
     * @return {?}
     */
    OidcSecurityService.prototype.createEndSessionUrl = function (end_session_endpoint, id_token_hint) {
        var /** @type {?} */ urlParts = end_session_endpoint.split('?');
        var /** @type {?} */ authorizationEndsessionUrl = urlParts[0];
        var /** @type {?} */ params = new HttpParams({
            fromString: urlParts[1],
            encoder: new UriEncoder()
        });
        params = params.set('id_token_hint', id_token_hint);
        params = params.append('post_logout_redirect_uri', this.authConfiguration.post_logout_redirect_uri);
        return authorizationEndsessionUrl + "?" + params;
    };
    /**
     * @param {?} isRenewProcess
     * @return {?}
     */
    OidcSecurityService.prototype.resetAuthorizationData = function (isRenewProcess) {
        if (!isRenewProcess) {
            if (this.authConfiguration.auto_userinfo) {
                // Clear user data. Fixes #97.
                this.setUserData('');
            }
            this.setIsAuthorized(false);
            this.oidcSecurityCommon.resetStorageData(isRenewProcess);
            this.checkSessionChanged = false;
        }
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.onUserDataChanged = function () {
        this.loggerService.logDebug("onUserDataChanged: last = " + this.lastUserData + ", new = " + this._userData.value);
        if (this.lastUserData && !this._userData.value) {
            this.loggerService.logDebug('onUserDataChanged: Logout detected.');
            // TODO should we have an action here
        }
        this.lastUserData = this._userData.value;
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.getSigningKeys = function () {
        if (this.authWellKnownEndpoints) {
            this.loggerService.logDebug('jwks_uri: ' + this.authWellKnownEndpoints.jwks_uri);
            return this.oidcDataService
                .get(this.authWellKnownEndpoints.jwks_uri)
                .pipe(catchError(this.handleErrorGetSigningKeys));
        }
        else {
            this.loggerService.logWarning('getSigningKeys: authWellKnownEndpoints is undefined');
        }
        return this.oidcDataService
            .get('undefined')
            .pipe(catchError(this.handleErrorGetSigningKeys));
    };
    /**
     * @param {?} error
     * @return {?}
     */
    OidcSecurityService.prototype.handleErrorGetSigningKeys = function (error) {
        var /** @type {?} */ errMsg;
        if (error instanceof Response) {
            var /** @type {?} */ body = error.json() || {};
            var /** @type {?} */ err = JSON.stringify(body);
            errMsg = error.status + " - " + (error.statusText || '') + " " + err;
        }
        else {
            errMsg = error.message ? error.message : error.toString();
        }
        console.error(errMsg);
        return throwError(errMsg);
    };
    /**
     * @return {?}
     */
    OidcSecurityService.prototype.runTokenValidation = function () {
        var _this = this;
        if (this.runTokenValidationRunning) {
            return;
        }
        this.runTokenValidationRunning = true;
        /**
         * First time: delay 10 seconds to call silentRenewHeartBeatCheck
         * Afterwards: Run this check in a 5 second interval only AFTER the previous operation ends.
         */
        var /** @type {?} */ silentRenewHeartBeatCheck = function () {
            if (_this._userData.value && (_this.oidcSecurityCommon.silentRenewRunning !== 'running') && _this.getIdToken()) {
                if (_this.oidcSecurityValidation.isTokenExpired(_this.oidcSecurityCommon.idToken, _this.authConfiguration.silent_renew_offset_in_seconds)) {
                    _this.loggerService.logDebug('IsAuthorized: id_token isTokenExpired, start silent renew if active');
                    if (_this.authConfiguration.silent_renew) {
                        _this.refreshSession().subscribe(function () {
                            _this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 3000);
                        }, function (err) {
                            _this.loggerService.logError('Error: ' + err);
                            _this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 3000);
                        });
                        /* In this situation, we schedule a heatbeat check only when silentRenew is finished.
                                                We don't want to schedule another check so we have to return here */
                        return;
                    }
                    else {
                        _this.resetAuthorizationData(false);
                    }
                }
            }
            /* Delay 3 seconds and do the next check */
            _this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 3000);
        };
        this.zone.runOutsideAngular(function () {
            /* Initial heartbeat check */
            _this._scheduledHeartBeat = setTimeout(silentRenewHeartBeatCheck, 10000);
        });
    };
    /**
     * @param {?} e
     * @return {?}
     */
    OidcSecurityService.prototype.silentRenewEventHandler = function (e) {
        this.loggerService.logDebug('silentRenewEventHandler');
        this.authorizedCallback(e.detail);
    };
    return OidcSecurityService;
}());
OidcSecurityService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
OidcSecurityService.ctorParameters = function () { return [
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
    { type: OidcDataService, },
    { type: StateValidationService, },
    { type: AuthConfiguration, },
    { type: Router, },
    { type: OidcSecurityCheckSession, },
    { type: OidcSecuritySilentRenew, },
    { type: OidcSecurityUserService, },
    { type: OidcSecurityCommon, },
    { type: OidcSecurityValidation, },
    { type: TokenHelperService, },
    { type: LoggerService, },
    { type: NgZone, },
]; };
OidcSecurityService.propDecorators = {
    "onModuleSetup": [{ type: Output },],
    "onAuthorizationResult": [{ type: Output },],
    "onCheckSessionChanged": [{ type: Output },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AuthWellKnownEndpoints = /** @class */ (function () {
    function AuthWellKnownEndpoints() {
        this.issuer = '';
        this.jwks_uri = '';
        this.authorization_endpoint = '';
        this.token_endpoint = '';
        this.userinfo_endpoint = '';
        this.end_session_endpoint = '';
        this.check_session_iframe = '';
        this.revocation_endpoint = '';
        this.introspection_endpoint = '';
    }
    /**
     * @param {?} data
     * @return {?}
     */
    AuthWellKnownEndpoints.prototype.setWellKnownEndpoints = function (data) {
        this.issuer = data.issuer;
        this.jwks_uri = data.jwks_uri;
        this.authorization_endpoint = data.authorization_endpoint;
        this.token_endpoint = data.token_endpoint;
        this.userinfo_endpoint = data.userinfo_endpoint;
        if (data.end_session_endpoint) {
            this.end_session_endpoint = data.end_session_endpoint;
        }
        if (data.check_session_iframe) {
            this.check_session_iframe = data.check_session_iframe;
        }
        if (data.revocation_endpoint) {
            this.revocation_endpoint = data.revocation_endpoint;
        }
        if (data.introspection_endpoint) {
            this.introspection_endpoint = data.introspection_endpoint;
        }
    };
    return AuthWellKnownEndpoints;
}());
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var JwtKeys = /** @class */ (function () {
    function JwtKeys() {
        this.keys = [];
    }
    return JwtKeys;
}());
var JwtKey = /** @class */ (function () {
    function JwtKey() {
        this.kty = '';
        this.use = '';
        this.kid = '';
        this.x5t = '';
        this.e = '';
        this.n = '';
        this.x5c = [];
    }
    return JwtKey;
}());
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AuthModule = /** @class */ (function () {
    function AuthModule() {
    }
    /**
     * @param {?=} token
     * @return {?}
     */
    AuthModule.forRoot = function (token) {
        if (token === void 0) { token = {}; }
        return {
            ngModule: AuthModule,
            providers: [
                OidcConfigService,
                OidcSecurityService,
                OidcSecurityValidation,
                OidcSecurityCheckSession,
                OidcSecuritySilentRenew,
                OidcSecurityUserService,
                OidcSecurityCommon,
                AuthConfiguration,
                TokenHelperService,
                LoggerService,
                DefaultConfiguration,
                ArrayHelperService,
                AuthWellKnownEndpoints,
                OidcDataService,
                StateValidationService,
                {
                    provide: OidcSecurityStorage,
                    useClass: token.storage || BrowserStorage
                }
            ]
        };
    };
    return AuthModule;
}());
AuthModule.decorators = [
    { type: NgModule },
];
/**
 * @record
 * @template T
 */
/**
 * @record
 */
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// Public classes.
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Angular OpenID Connect Implicit Flow.
 * OpenID Connect Implicit Flow RP Client
 * Written by Damien Bowden.
 * MIT license.
 * https://github.com/damienbod/angular-auth-oidc-client
 */
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { OidcConfigService, OidcSecurityService, OidcSecurityStorage, BrowserStorage, AuthWellKnownEndpoints, AuthorizationResult, JwtKeys, JwtKey, ValidateStateResult, DefaultConfiguration, OpenIDImplicitFlowConfiguration, AuthConfiguration, AuthModule, OidcSecurityValidation, ArrayHelperService as d, OidcDataService as a, StateValidationService as b, TokenHelperService as e, LoggerService as f, OidcSecurityCheckSession as g, OidcSecurityCommon as c, OidcSecuritySilentRenew as h, OidcSecurityUserService as i };
